/*
  File autogenerated by gengetopt version 2.22.5
  generated with the following command:
  gengetopt -i sources/include/xhmmCmdline.ggo -F xhmmCmdline --header-output-dir=./sources/include --src-output-dir=./sources/lib --conf-parser 

  The developers of gengetopt consider the fixed text that goes in all
  gengetopt output files to be in the public domain:
  we make no copyright claims on it.
*/

/* If we use autoconf.  */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef FIX_UNUSED
#define FIX_UNUSED(X) (void) (X) /* avoid warnings for unused params */
#endif

#include <getopt.h>

#include "xhmmCmdline.h"

const char *gengetopt_args_info_purpose = "Uses principal component analysis (PCA) normalization and a hidden Markov model \n(HMM) to detect and genotype copy number variation (CNV) from normalized \nread-depth data from targeted sequencing experiments.";

const char *gengetopt_args_info_usage = "Usage: xhmm [OPTIONS]...";

const char *gengetopt_args_info_description = "Developed by Menachem Fromer and Shaun Purcell";

const char *gengetopt_args_info_detailed_help[] = {
  "  -h, --help                    Print help and exit",
  "      --detailed-help           Print help, including all details and hidden \n                                  options, and exit",
  "      --full-help               Print help, including hidden options, and exit",
  "  -V, --version                 Print version and exit",
  "\n*******************************************************************\nOptions for modes: 'prepareTargets', 'genotype':",
  "  -F, --referenceFASTA=STRING   Reference FASTA file (MUST have .fai index \n                                  file)",
  "\n*******************************************************************\nOptions for modes: 'matrix', 'PCA', 'normalize', 'discover', 'genotype':",
  "  -r, --readDepths=STRING       Matrix of *input* read-depths, where rows \n                                  (samples) and columns (targets) are labeled  \n                                  (default=`-')",
  "  ",
  "*******************************************************************",
  "\n Mode: prepareTargets\n  Sort all target intervals, merge overlapping ones, and print the resulting \n  interval list",
  "  -S, --prepareTargets          ",
  "      --targets=STRING          Input targets lists",
  "      --mergedTargets=STRING    Output targets list  (default=`-')",
  "\n Mode: mergeGATKdepths\n  Merge the output from GATK into a single read depth matrix of samples (rows) \n  by targets (columns)",
  "  -A, --mergeGATKdepths         ",
  "      --GATKdepths=STRING       GATK sample_interval_summary output file(s) to \n                                  be merged [must have *IDENTICAL* target \n                                  lists]",
  "      --GATKdepthsList=STRING   A file containing a list of GATK \n                                  sample_interval_summary output files to be \n                                  merged [must have *IDENTICAL* target lists]",
  "      --sampleIDmap=STRING      File containing mappings of sample names to new \n                                  sample names (in columns designated by \n                                  fromID, toID)",
  "      --fromID=INT              Column number of OLD sample IDs to map  \n                                  (default=`1')",
  "      --toID=INT                Column number of NEW sample IDs to map  \n                                  (default=`2')",
  "      --columnSuffix=STRING     Suffix of columns to be used for merging [where \n                                  columns are in the form: SAMPLE + \n                                  columnSuffix]  (default=`_mean_cvg')",
  "      --rdPrecision=INT         Decimal precision of read depths output  \n                                  (default=`2')",
  "      --outputTargetsBySamples  Output targets x samples (instead of samples x \n                                  targets)  (default=off)",
  "\n Mode: matrix\n  Process (filter, center, etc.) a read depth matrix and output the resulting \n  matrix.  Note that first all excluded samples and targets are removed.  And, \n  sample statistics used for filtering are calculated only *after* filtering \n  out relevant targets.",
  "  -M, --matrix                  ",
  "      --excludeTargets=STRING   File(s) of targets to exclude",
  "      --excludeChromosomeTargets=STRING\n                                Target chromosome(s) to exclude",
  "      --excludeSamples=STRING   File(s) of samples to exclude",
  "      --minTargetSize=INT       Minimum size of target (in bp) to process  \n                                  (default=`0')",
  "      --maxTargetSize=INT       Maximum size of target (in bp) to process",
  "      --minMeanTargetRD=DOUBLE  Minimum per-target mean RD to require for \n                                  target to be processed",
  "      --maxMeanTargetRD=DOUBLE  Maximum per-target mean RD to require for \n                                  target to be processed",
  "      --minSdTargetRD=DOUBLE    Minimum per-target standard deviation of RD to \n                                  require for target to be processed  \n                                  (default=`0')",
  "      --maxSdTargetRD=DOUBLE    Maximum per-target standard deviation of RD to \n                                  require for target to be processed",
  "      --minMeanSampleRD=DOUBLE  Minimum per-sample mean RD to require for \n                                  sample to be processed",
  "      --maxMeanSampleRD=DOUBLE  Maximum per-sample mean RD to require for \n                                  sample to be processed",
  "      --minSdSampleRD=DOUBLE    Minimum per-sample standard deviation of RD to \n                                  require for sample to be processed  \n                                  (default=`0')",
  "      --maxSdSampleRD=DOUBLE    Maximum per-sample standard deviation of RD to \n                                  require for sample to be processed",
  "      --scaleDataBySum          After any filtering, scale read-depth matrix \n                                  values by sample- or target- sums (as per \n                                  --scaleDataBySumType) [i.e., divide by row or \n                                  column sums], but multiply by factor \n                                  specificied by --scaleDataBySumFactor  \n                                  (default=off)",
  "      --scaleDataBySumType=ENUM If --scaleDataBySum given, then scale the data \n                                  within this dimension  (possible \n                                  values=\"target\", \"sample\")",
  "      --scaleDataBySumFactor=DOUBLE\n                                If --scaleDataBySum given, then divide by \n                                  appropriate sum (but multiply by this factor) \n                                   (default=`1e6')",
  "      --log10=DOUBLE            After any filtering and optional scaling steps \n                                  (but before any optional centering steps), \n                                  convert the matrix to log10 values. To deal \n                                  with non-positive and small positive values, \n                                  a truncation and then pseudocount approach is \n                                  taken. Specifically, denote the original \n                                  matrix value as x and this parameter's \n                                  pseudocount value as v (say, 0.5). The matrix \n                                  value used is then log10(max(x, 0) + v)",
  "      --centerData              Output sample- or target- centered read-depth \n                                  matrix (as per --centerType)  (default=off)",
  "      --centerType=ENUM         If --centerData given, then center the data \n                                  around this dimension  (possible \n                                  values=\"target\", \"sample\")",
  "      --zScoreData              If --centerData given, then additionally \n                                  normalize by standard deviation (outputting \n                                  z-scores)  (default=off)",
  "      --outputExcludedTargets=STRING\n                                File in which to output targets excluded by \n                                  some criterion",
  "      --outputExcludedSamples=STRING\n                                File in which to output samples excluded by \n                                  some criterion",
  "\nOptions for modes: 'mergeGATKdepths', 'matrix':",
  "  -o, --outputMatrix=STRING     Read-depth matrix output file  (default=`-')",
  "*******************************************************************",
  "\n Mode: PCA\n  Run PCA to create normalization factors for read depth matrix",
  "  -P, --PCA                     Matrix is read from --readDepths argument; \n                                  normalization factors sent to --PCAfiles \n                                  argument",
  "      --PCA_saveMemory[=STRING] Should XHMM save memory by storing some of the \n                                  intermediate PCA matrices as temporary files \n                                  on disk?  (default=`')",
  "  Note that, if no argument is given (syntax is '--PCA_saveMemory=TEMP_DIR' \n  instead of just '--PCA_saveMemory'), the temporary files will be generated in \n  the same directory as specified by the --PCAfiles argument",
  "\n Mode: normalize\n  Apply PCA factors in order to normalize read depth matrix",
  "  -N, --normalize               Matrix is read from --readDepths argument; \n                                  normalization factors read from --PCAfiles \n                                  argument",
  "  -n, --normalizeOutput=STRING  Normalized read-depth matrix output file  \n                                  (default=`-')",
  "      --PCnormalizeMethod=ENUM  Method to choose which prinicipal components \n                                  are removed for data normalization  (possible \n                                  values=\"numPCtoRemove\", \"PVE_mean\", \n                                  \"PVE_contrib\" default=`PVE_mean')",
  "      --numPCtoRemove=INT       Number of highest principal components to \n                                  filter out  (default=`20')",
  "      --PVE_mean_factor=DOUBLE  Remove all principal components that \n                                  individually explain more variance than this \n                                  factor times the average (in the original \n                                  PCA-ed data)  (default=`0.7')",
  "      --PVE_contrib=DOUBLE      Remove the smallest number of principal \n                                  components that explain this percent of the \n                                  variance (in the original PCA-ed data)  \n                                  (default=`50')",
  "\nOptions for modes: 'PCA', 'normalize':",
  "      --PCAfiles=STRING         Base file name for 'PCA' *output*, and \n                                  'normalize' *input*",
  "  This prefix will name 3 files: the principal components (PC), the sample \n  loadings, and the standard deviations in each PC.",
  "  -C, --createDB                ",
  "*******************************************************************",
  "\n Mode: discover\n  Discover CNVs from normalized read depth matrix",
  "  -D, --discover                Matrix is read from --readDepths argument",
  "  -c, --xcnv=STRING             CNV output file  (default=`-')",
  "      --optDiscover             Optimize input HMM parameters before \n                                  discovering CNVs",
  "  -t, --discoverSomeQualThresh=DOUBLE\n                                Quality threshold for discovering a CNV in a \n                                  sample  (default=`30')",
  "  -s, --posteriorFiles=STRING   Base file name for posterior probabilities \n                                  output files; if not given, and --xcnv is not \n                                  '-', this will default to --xcnv argument",
  "\n Mode: genotype\n  Genotype list of CNVs from normalized read depth matrix",
  "  -G, --genotype                Matrix is read from --readDepths argument",
  "  -g, --gxcnv=STRING            xhmm CNV input file to genotype in 'readDepths' \n                                  sample",
  "      --subsegments             In addition to genotyping the intervals \n                                  specified in gxcnv, genotype all sub-segments \n                                  of these intervals (with \n                                  maxTargetsInSubsegment or fewer targets)  \n                                  (default=off)",
  "  Note that the genotyping threshold is set to \n  max(genotypeQualThresholdWhenNoExact, GQT for original event)",
  "      --maxTargetsInSubsegment=INT\n                                When genotyping sub-segments of input \n                                  intervals, only consider sub-segments \n                                  consisting of this number of targets or fewer \n                                   (default=`30')",
  "  -T, --genotypeQualThresholdWhenNoExact=DOUBLE\n                                Quality threshold for calling a genotype, used \n                                  *ONLY* when 'gxcnv' does not contain the \n                                  'Q_EXACT' field for the interval being \n                                  genotyped  (default=`20')",
  "  When the 'gxcnv' file contains the 'Q_EXACT' field, then the minimal such \n  value for any called CNV is used as the genotyping threshold for that CNV.  \n  Otherwise, the value given here is used.",
  "\n Mode: mergeVCFs\n  Merge VCF files output by multiple runs of the 'genotype' command on the same \n  input intervals (.xcnv file), but for different samples",
  "      --mergeVCFs               ",
  "      --mergeVCF=STRING         VCF file(s) to be merged [must have *IDENTICAL* \n                                  genotyped intervals]",
  "      --mergeVCFlist=STRING     A file containing a list of VCF files to be \n                                  merged [must have *IDENTICAL* genotyped \n                                  intervals]",
  "\nOptions for modes: 'discover', 'genotype', 'transition', 'printHMM':",
  "  -p, --paramFile=STRING        (Initial) model parameters file",
  "  File consisting of fields:  1e-8	6	70	-3	1.00	0	1.00	3	1.00",
  "\nOptions for modes: 'discover', 'genotype':",
  "  -d, --DB=STRING               Database storage file",
  "  If given for 'discover', 'genotype', or 'createDB', the 'readDepths' data and \n  'discover' results will be cached to the DB for long-term storage",
  "  -m, --maxNormalizedReadDepthVal=DOUBLE\n                                Value at which to cap the absolute value of \n                                  *normalized* input read depth values \n                                  ('readDepths')  (default=`10')",
  "  -q, --maxQualScore=DOUBLE     Value at which to cap the calculated quality \n                                  scores  (default=`99')",
  "  -e, --scorePrecision=INT      Decimal precision of quality scores  \n                                  (default=`0')",
  "  -a, --aux_xcnv=STRING         Auxiliary CNV output file (may be VERY LARGE in \n                                  'genotype' mode)",
  "  Each line gives detailed information regarding each target that is part of a \n  CNV in the output file",
  "  -u, --auxUpstreamPrintTargs=INT\n                                Number of targets to print upstream of CNV in \n                                  'auxOutput' file  (default=`2')",
  "  -w, --auxDownstreamPrintTargs=INT\n                                Number of targets to print downstream of CNV in \n                                  'auxOutput' file  (default=`2')",
  "  -R, --origReadDepths=STRING   Matrix of unnormalized read-depths to use for \n                                  CNV annotation, where rows (samples) and \n                                  columns (targets) are labeled",
  "  These values are used to annotate the CNV output.",
  "      --keepSampleIDs=STRING    File containing a list of samples to be \n                                  analyzed",
  "  If this parameter is given, any sample not listed in this file will be \n  dropped from this run.",
  "\nOptions for modes: 'genotype', 'mergeVCFs':",
  "  -v, --vcf=STRING              Genotyped CNV output VCF file  (default=`-')",
  "*******************************************************************",
  "\n Mode: printHMM\n  Print HMM model parameters and exit",
  "      --printHMM                ",
  "\n Mode: transition\n  Print HMM transition matrix for user-requested genomic distances",
  "      --transition              ",
    0
};
static void
init_full_help_array(void)
{
  gengetopt_args_info_full_help[0] = gengetopt_args_info_detailed_help[0];
  gengetopt_args_info_full_help[1] = gengetopt_args_info_detailed_help[1];
  gengetopt_args_info_full_help[2] = gengetopt_args_info_detailed_help[2];
  gengetopt_args_info_full_help[3] = gengetopt_args_info_detailed_help[3];
  gengetopt_args_info_full_help[4] = gengetopt_args_info_detailed_help[4];
  gengetopt_args_info_full_help[5] = gengetopt_args_info_detailed_help[5];
  gengetopt_args_info_full_help[6] = gengetopt_args_info_detailed_help[6];
  gengetopt_args_info_full_help[7] = gengetopt_args_info_detailed_help[7];
  gengetopt_args_info_full_help[8] = gengetopt_args_info_detailed_help[9];
  gengetopt_args_info_full_help[9] = gengetopt_args_info_detailed_help[10];
  gengetopt_args_info_full_help[10] = gengetopt_args_info_detailed_help[11];
  gengetopt_args_info_full_help[11] = gengetopt_args_info_detailed_help[12];
  gengetopt_args_info_full_help[12] = gengetopt_args_info_detailed_help[13];
  gengetopt_args_info_full_help[13] = gengetopt_args_info_detailed_help[14];
  gengetopt_args_info_full_help[14] = gengetopt_args_info_detailed_help[15];
  gengetopt_args_info_full_help[15] = gengetopt_args_info_detailed_help[16];
  gengetopt_args_info_full_help[16] = gengetopt_args_info_detailed_help[17];
  gengetopt_args_info_full_help[17] = gengetopt_args_info_detailed_help[18];
  gengetopt_args_info_full_help[18] = gengetopt_args_info_detailed_help[19];
  gengetopt_args_info_full_help[19] = gengetopt_args_info_detailed_help[20];
  gengetopt_args_info_full_help[20] = gengetopt_args_info_detailed_help[21];
  gengetopt_args_info_full_help[21] = gengetopt_args_info_detailed_help[22];
  gengetopt_args_info_full_help[22] = gengetopt_args_info_detailed_help[23];
  gengetopt_args_info_full_help[23] = gengetopt_args_info_detailed_help[24];
  gengetopt_args_info_full_help[24] = gengetopt_args_info_detailed_help[25];
  gengetopt_args_info_full_help[25] = gengetopt_args_info_detailed_help[26];
  gengetopt_args_info_full_help[26] = gengetopt_args_info_detailed_help[27];
  gengetopt_args_info_full_help[27] = gengetopt_args_info_detailed_help[28];
  gengetopt_args_info_full_help[28] = gengetopt_args_info_detailed_help[29];
  gengetopt_args_info_full_help[29] = gengetopt_args_info_detailed_help[30];
  gengetopt_args_info_full_help[30] = gengetopt_args_info_detailed_help[31];
  gengetopt_args_info_full_help[31] = gengetopt_args_info_detailed_help[32];
  gengetopt_args_info_full_help[32] = gengetopt_args_info_detailed_help[33];
  gengetopt_args_info_full_help[33] = gengetopt_args_info_detailed_help[34];
  gengetopt_args_info_full_help[34] = gengetopt_args_info_detailed_help[35];
  gengetopt_args_info_full_help[35] = gengetopt_args_info_detailed_help[36];
  gengetopt_args_info_full_help[36] = gengetopt_args_info_detailed_help[37];
  gengetopt_args_info_full_help[37] = gengetopt_args_info_detailed_help[38];
  gengetopt_args_info_full_help[38] = gengetopt_args_info_detailed_help[39];
  gengetopt_args_info_full_help[39] = gengetopt_args_info_detailed_help[40];
  gengetopt_args_info_full_help[40] = gengetopt_args_info_detailed_help[41];
  gengetopt_args_info_full_help[41] = gengetopt_args_info_detailed_help[42];
  gengetopt_args_info_full_help[42] = gengetopt_args_info_detailed_help[43];
  gengetopt_args_info_full_help[43] = gengetopt_args_info_detailed_help[44];
  gengetopt_args_info_full_help[44] = gengetopt_args_info_detailed_help[45];
  gengetopt_args_info_full_help[45] = gengetopt_args_info_detailed_help[46];
  gengetopt_args_info_full_help[46] = gengetopt_args_info_detailed_help[47];
  gengetopt_args_info_full_help[47] = gengetopt_args_info_detailed_help[48];
  gengetopt_args_info_full_help[48] = gengetopt_args_info_detailed_help[49];
  gengetopt_args_info_full_help[49] = gengetopt_args_info_detailed_help[50];
  gengetopt_args_info_full_help[50] = gengetopt_args_info_detailed_help[51];
  gengetopt_args_info_full_help[51] = gengetopt_args_info_detailed_help[52];
  gengetopt_args_info_full_help[52] = gengetopt_args_info_detailed_help[53];
  gengetopt_args_info_full_help[53] = gengetopt_args_info_detailed_help[55];
  gengetopt_args_info_full_help[54] = gengetopt_args_info_detailed_help[56];
  gengetopt_args_info_full_help[55] = gengetopt_args_info_detailed_help[57];
  gengetopt_args_info_full_help[56] = gengetopt_args_info_detailed_help[58];
  gengetopt_args_info_full_help[57] = gengetopt_args_info_detailed_help[59];
  gengetopt_args_info_full_help[58] = gengetopt_args_info_detailed_help[60];
  gengetopt_args_info_full_help[59] = gengetopt_args_info_detailed_help[61];
  gengetopt_args_info_full_help[60] = gengetopt_args_info_detailed_help[62];
  gengetopt_args_info_full_help[61] = gengetopt_args_info_detailed_help[63];
  gengetopt_args_info_full_help[62] = gengetopt_args_info_detailed_help[65];
  gengetopt_args_info_full_help[63] = gengetopt_args_info_detailed_help[66];
  gengetopt_args_info_full_help[64] = gengetopt_args_info_detailed_help[67];
  gengetopt_args_info_full_help[65] = gengetopt_args_info_detailed_help[68];
  gengetopt_args_info_full_help[66] = gengetopt_args_info_detailed_help[69];
  gengetopt_args_info_full_help[67] = gengetopt_args_info_detailed_help[70];
  gengetopt_args_info_full_help[68] = gengetopt_args_info_detailed_help[71];
  gengetopt_args_info_full_help[69] = gengetopt_args_info_detailed_help[72];
  gengetopt_args_info_full_help[70] = gengetopt_args_info_detailed_help[73];
  gengetopt_args_info_full_help[71] = gengetopt_args_info_detailed_help[74];
  gengetopt_args_info_full_help[72] = gengetopt_args_info_detailed_help[75];
  gengetopt_args_info_full_help[73] = gengetopt_args_info_detailed_help[76];
  gengetopt_args_info_full_help[74] = gengetopt_args_info_detailed_help[78];
  gengetopt_args_info_full_help[75] = gengetopt_args_info_detailed_help[79];
  gengetopt_args_info_full_help[76] = gengetopt_args_info_detailed_help[81];
  gengetopt_args_info_full_help[77] = gengetopt_args_info_detailed_help[82];
  gengetopt_args_info_full_help[78] = gengetopt_args_info_detailed_help[83];
  gengetopt_args_info_full_help[79] = gengetopt_args_info_detailed_help[84];
  gengetopt_args_info_full_help[80] = gengetopt_args_info_detailed_help[85];
  gengetopt_args_info_full_help[81] = gengetopt_args_info_detailed_help[86];
  gengetopt_args_info_full_help[82] = gengetopt_args_info_detailed_help[88];
  gengetopt_args_info_full_help[83] = gengetopt_args_info_detailed_help[89];
  gengetopt_args_info_full_help[84] = gengetopt_args_info_detailed_help[91];
  gengetopt_args_info_full_help[85] = gengetopt_args_info_detailed_help[92];
  gengetopt_args_info_full_help[86] = gengetopt_args_info_detailed_help[93];
  gengetopt_args_info_full_help[87] = gengetopt_args_info_detailed_help[94];
  gengetopt_args_info_full_help[88] = gengetopt_args_info_detailed_help[96];
  gengetopt_args_info_full_help[89] = gengetopt_args_info_detailed_help[97];
  gengetopt_args_info_full_help[90] = gengetopt_args_info_detailed_help[98];
  gengetopt_args_info_full_help[91] = gengetopt_args_info_detailed_help[100];
  gengetopt_args_info_full_help[92] = gengetopt_args_info_detailed_help[102];
  gengetopt_args_info_full_help[93] = gengetopt_args_info_detailed_help[103];
  gengetopt_args_info_full_help[94] = gengetopt_args_info_detailed_help[104];
  gengetopt_args_info_full_help[95] = gengetopt_args_info_detailed_help[105];
  gengetopt_args_info_full_help[96] = gengetopt_args_info_detailed_help[106];
  gengetopt_args_info_full_help[97] = gengetopt_args_info_detailed_help[107];
  gengetopt_args_info_full_help[98] = gengetopt_args_info_detailed_help[108];
  gengetopt_args_info_full_help[99] = 0; 
  
}

const char *gengetopt_args_info_full_help[100];

static void
init_help_array(void)
{
  gengetopt_args_info_help[0] = gengetopt_args_info_detailed_help[0];
  gengetopt_args_info_help[1] = gengetopt_args_info_detailed_help[1];
  gengetopt_args_info_help[2] = gengetopt_args_info_detailed_help[2];
  gengetopt_args_info_help[3] = gengetopt_args_info_detailed_help[3];
  gengetopt_args_info_help[4] = gengetopt_args_info_detailed_help[4];
  gengetopt_args_info_help[5] = gengetopt_args_info_detailed_help[5];
  gengetopt_args_info_help[6] = gengetopt_args_info_detailed_help[6];
  gengetopt_args_info_help[7] = gengetopt_args_info_detailed_help[7];
  gengetopt_args_info_help[8] = gengetopt_args_info_detailed_help[9];
  gengetopt_args_info_help[9] = gengetopt_args_info_detailed_help[10];
  gengetopt_args_info_help[10] = gengetopt_args_info_detailed_help[11];
  gengetopt_args_info_help[11] = gengetopt_args_info_detailed_help[12];
  gengetopt_args_info_help[12] = gengetopt_args_info_detailed_help[13];
  gengetopt_args_info_help[13] = gengetopt_args_info_detailed_help[14];
  gengetopt_args_info_help[14] = gengetopt_args_info_detailed_help[15];
  gengetopt_args_info_help[15] = gengetopt_args_info_detailed_help[16];
  gengetopt_args_info_help[16] = gengetopt_args_info_detailed_help[17];
  gengetopt_args_info_help[17] = gengetopt_args_info_detailed_help[18];
  gengetopt_args_info_help[18] = gengetopt_args_info_detailed_help[19];
  gengetopt_args_info_help[19] = gengetopt_args_info_detailed_help[20];
  gengetopt_args_info_help[20] = gengetopt_args_info_detailed_help[21];
  gengetopt_args_info_help[21] = gengetopt_args_info_detailed_help[22];
  gengetopt_args_info_help[22] = gengetopt_args_info_detailed_help[23];
  gengetopt_args_info_help[23] = gengetopt_args_info_detailed_help[24];
  gengetopt_args_info_help[24] = gengetopt_args_info_detailed_help[25];
  gengetopt_args_info_help[25] = gengetopt_args_info_detailed_help[26];
  gengetopt_args_info_help[26] = gengetopt_args_info_detailed_help[27];
  gengetopt_args_info_help[27] = gengetopt_args_info_detailed_help[28];
  gengetopt_args_info_help[28] = gengetopt_args_info_detailed_help[29];
  gengetopt_args_info_help[29] = gengetopt_args_info_detailed_help[30];
  gengetopt_args_info_help[30] = gengetopt_args_info_detailed_help[31];
  gengetopt_args_info_help[31] = gengetopt_args_info_detailed_help[32];
  gengetopt_args_info_help[32] = gengetopt_args_info_detailed_help[33];
  gengetopt_args_info_help[33] = gengetopt_args_info_detailed_help[34];
  gengetopt_args_info_help[34] = gengetopt_args_info_detailed_help[35];
  gengetopt_args_info_help[35] = gengetopt_args_info_detailed_help[36];
  gengetopt_args_info_help[36] = gengetopt_args_info_detailed_help[37];
  gengetopt_args_info_help[37] = gengetopt_args_info_detailed_help[38];
  gengetopt_args_info_help[38] = gengetopt_args_info_detailed_help[39];
  gengetopt_args_info_help[39] = gengetopt_args_info_detailed_help[40];
  gengetopt_args_info_help[40] = gengetopt_args_info_detailed_help[41];
  gengetopt_args_info_help[41] = gengetopt_args_info_detailed_help[42];
  gengetopt_args_info_help[42] = gengetopt_args_info_detailed_help[43];
  gengetopt_args_info_help[43] = gengetopt_args_info_detailed_help[44];
  gengetopt_args_info_help[44] = gengetopt_args_info_detailed_help[45];
  gengetopt_args_info_help[45] = gengetopt_args_info_detailed_help[46];
  gengetopt_args_info_help[46] = gengetopt_args_info_detailed_help[47];
  gengetopt_args_info_help[47] = gengetopt_args_info_detailed_help[48];
  gengetopt_args_info_help[48] = gengetopt_args_info_detailed_help[49];
  gengetopt_args_info_help[49] = gengetopt_args_info_detailed_help[50];
  gengetopt_args_info_help[50] = gengetopt_args_info_detailed_help[51];
  gengetopt_args_info_help[51] = gengetopt_args_info_detailed_help[52];
  gengetopt_args_info_help[52] = gengetopt_args_info_detailed_help[53];
  gengetopt_args_info_help[53] = gengetopt_args_info_detailed_help[55];
  gengetopt_args_info_help[54] = gengetopt_args_info_detailed_help[56];
  gengetopt_args_info_help[55] = gengetopt_args_info_detailed_help[57];
  gengetopt_args_info_help[56] = gengetopt_args_info_detailed_help[58];
  gengetopt_args_info_help[57] = gengetopt_args_info_detailed_help[59];
  gengetopt_args_info_help[58] = gengetopt_args_info_detailed_help[60];
  gengetopt_args_info_help[59] = gengetopt_args_info_detailed_help[61];
  gengetopt_args_info_help[60] = gengetopt_args_info_detailed_help[62];
  gengetopt_args_info_help[61] = gengetopt_args_info_detailed_help[63];
  gengetopt_args_info_help[62] = gengetopt_args_info_detailed_help[66];
  gengetopt_args_info_help[63] = gengetopt_args_info_detailed_help[67];
  gengetopt_args_info_help[64] = gengetopt_args_info_detailed_help[68];
  gengetopt_args_info_help[65] = gengetopt_args_info_detailed_help[69];
  gengetopt_args_info_help[66] = gengetopt_args_info_detailed_help[71];
  gengetopt_args_info_help[67] = gengetopt_args_info_detailed_help[72];
  gengetopt_args_info_help[68] = gengetopt_args_info_detailed_help[73];
  gengetopt_args_info_help[69] = gengetopt_args_info_detailed_help[74];
  gengetopt_args_info_help[70] = gengetopt_args_info_detailed_help[75];
  gengetopt_args_info_help[71] = gengetopt_args_info_detailed_help[76];
  gengetopt_args_info_help[72] = gengetopt_args_info_detailed_help[78];
  gengetopt_args_info_help[73] = gengetopt_args_info_detailed_help[79];
  gengetopt_args_info_help[74] = gengetopt_args_info_detailed_help[81];
  gengetopt_args_info_help[75] = gengetopt_args_info_detailed_help[82];
  gengetopt_args_info_help[76] = gengetopt_args_info_detailed_help[83];
  gengetopt_args_info_help[77] = gengetopt_args_info_detailed_help[84];
  gengetopt_args_info_help[78] = gengetopt_args_info_detailed_help[85];
  gengetopt_args_info_help[79] = gengetopt_args_info_detailed_help[86];
  gengetopt_args_info_help[80] = gengetopt_args_info_detailed_help[88];
  gengetopt_args_info_help[81] = gengetopt_args_info_detailed_help[91];
  gengetopt_args_info_help[82] = gengetopt_args_info_detailed_help[92];
  gengetopt_args_info_help[83] = gengetopt_args_info_detailed_help[93];
  gengetopt_args_info_help[84] = gengetopt_args_info_detailed_help[94];
  gengetopt_args_info_help[85] = gengetopt_args_info_detailed_help[96];
  gengetopt_args_info_help[86] = gengetopt_args_info_detailed_help[97];
  gengetopt_args_info_help[87] = gengetopt_args_info_detailed_help[98];
  gengetopt_args_info_help[88] = gengetopt_args_info_detailed_help[100];
  gengetopt_args_info_help[89] = gengetopt_args_info_detailed_help[102];
  gengetopt_args_info_help[90] = gengetopt_args_info_detailed_help[103];
  gengetopt_args_info_help[91] = gengetopt_args_info_detailed_help[104];
  gengetopt_args_info_help[92] = gengetopt_args_info_detailed_help[105];
  gengetopt_args_info_help[93] = gengetopt_args_info_detailed_help[106];
  gengetopt_args_info_help[94] = gengetopt_args_info_detailed_help[107];
  gengetopt_args_info_help[95] = gengetopt_args_info_detailed_help[108];
  gengetopt_args_info_help[96] = 0; 
  
}

const char *gengetopt_args_info_help[97];

typedef enum {ARG_NO
  , ARG_FLAG
  , ARG_STRING
  , ARG_INT
  , ARG_DOUBLE
  , ARG_ENUM
} cmdline_parser_arg_type;

static
void clear_given (struct gengetopt_args_info *args_info);
static
void clear_args (struct gengetopt_args_info *args_info);

static int
cmdline_parser_internal (int argc, char **argv, struct gengetopt_args_info *args_info,
                        struct cmdline_parser_params *params, const char *additional_error);

static int
cmdline_parser_required2 (struct gengetopt_args_info *args_info, const char *prog_name, const char *additional_error);
struct line_list
{
  char * string_arg;
  struct line_list * next;
};

static struct line_list *cmd_line_list = 0;
static struct line_list *cmd_line_list_tmp = 0;

static void
free_cmd_list(void)
{
  /* free the list of a previous call */
  if (cmd_line_list)
    {
      while (cmd_line_list) {
        cmd_line_list_tmp = cmd_line_list;
        cmd_line_list = cmd_line_list->next;
        free (cmd_line_list_tmp->string_arg);
        free (cmd_line_list_tmp);
      }
    }
}


const char *cmdline_parser_scaleDataBySumType_values[] = {"target", "sample", 0}; /*< Possible values for scaleDataBySumType. */
const char *cmdline_parser_centerType_values[] = {"target", "sample", 0}; /*< Possible values for centerType. */
const char *cmdline_parser_PCnormalizeMethod_values[] = {"numPCtoRemove", "PVE_mean", "PVE_contrib", 0}; /*< Possible values for PCnormalizeMethod. */

static char *
gengetopt_strdup (const char *s);

static
void clear_given (struct gengetopt_args_info *args_info)
{
  args_info->help_given = 0 ;
  args_info->detailed_help_given = 0 ;
  args_info->full_help_given = 0 ;
  args_info->version_given = 0 ;
  args_info->referenceFASTA_given = 0 ;
  args_info->readDepths_given = 0 ;
  args_info->prepareTargets_given = 0 ;
  args_info->targets_given = 0 ;
  args_info->mergedTargets_given = 0 ;
  args_info->mergeGATKdepths_given = 0 ;
  args_info->GATKdepths_given = 0 ;
  args_info->GATKdepthsList_given = 0 ;
  args_info->sampleIDmap_given = 0 ;
  args_info->fromID_given = 0 ;
  args_info->toID_given = 0 ;
  args_info->columnSuffix_given = 0 ;
  args_info->rdPrecision_given = 0 ;
  args_info->outputTargetsBySamples_given = 0 ;
  args_info->matrix_given = 0 ;
  args_info->excludeTargets_given = 0 ;
  args_info->excludeChromosomeTargets_given = 0 ;
  args_info->excludeSamples_given = 0 ;
  args_info->minTargetSize_given = 0 ;
  args_info->maxTargetSize_given = 0 ;
  args_info->minMeanTargetRD_given = 0 ;
  args_info->maxMeanTargetRD_given = 0 ;
  args_info->minSdTargetRD_given = 0 ;
  args_info->maxSdTargetRD_given = 0 ;
  args_info->minMeanSampleRD_given = 0 ;
  args_info->maxMeanSampleRD_given = 0 ;
  args_info->minSdSampleRD_given = 0 ;
  args_info->maxSdSampleRD_given = 0 ;
  args_info->scaleDataBySum_given = 0 ;
  args_info->scaleDataBySumType_given = 0 ;
  args_info->scaleDataBySumFactor_given = 0 ;
  args_info->log10_given = 0 ;
  args_info->centerData_given = 0 ;
  args_info->centerType_given = 0 ;
  args_info->zScoreData_given = 0 ;
  args_info->outputExcludedTargets_given = 0 ;
  args_info->outputExcludedSamples_given = 0 ;
  args_info->outputMatrix_given = 0 ;
  args_info->PCA_given = 0 ;
  args_info->PCA_saveMemory_given = 0 ;
  args_info->normalize_given = 0 ;
  args_info->normalizeOutput_given = 0 ;
  args_info->PCnormalizeMethod_given = 0 ;
  args_info->numPCtoRemove_given = 0 ;
  args_info->PVE_mean_factor_given = 0 ;
  args_info->PVE_contrib_given = 0 ;
  args_info->PCAfiles_given = 0 ;
  args_info->createDB_given = 0 ;
  args_info->discover_given = 0 ;
  args_info->xcnv_given = 0 ;
  args_info->optDiscover_given = 0 ;
  args_info->discoverSomeQualThresh_given = 0 ;
  args_info->posteriorFiles_given = 0 ;
  args_info->genotype_given = 0 ;
  args_info->gxcnv_given = 0 ;
  args_info->subsegments_given = 0 ;
  args_info->maxTargetsInSubsegment_given = 0 ;
  args_info->genotypeQualThresholdWhenNoExact_given = 0 ;
  args_info->mergeVCFs_given = 0 ;
  args_info->mergeVCF_given = 0 ;
  args_info->mergeVCFlist_given = 0 ;
  args_info->paramFile_given = 0 ;
  args_info->DB_given = 0 ;
  args_info->maxNormalizedReadDepthVal_given = 0 ;
  args_info->maxQualScore_given = 0 ;
  args_info->scorePrecision_given = 0 ;
  args_info->aux_xcnv_given = 0 ;
  args_info->auxUpstreamPrintTargs_given = 0 ;
  args_info->auxDownstreamPrintTargs_given = 0 ;
  args_info->origReadDepths_given = 0 ;
  args_info->keepSampleIDs_given = 0 ;
  args_info->vcf_given = 0 ;
  args_info->printHMM_given = 0 ;
  args_info->transition_given = 0 ;
  args_info->PCA_mode_counter = 0 ;
  args_info->discover_mode_counter = 0 ;
  args_info->genotype_mode_counter = 0 ;
  args_info->matrix_mode_counter = 0 ;
  args_info->mergeGATKdepths_mode_counter = 0 ;
  args_info->mergeVCFs_mode_counter = 0 ;
  args_info->normalize_mode_counter = 0 ;
  args_info->prepareTargets_mode_counter = 0 ;
  args_info->printHMM_mode_counter = 0 ;
  args_info->transition_mode_counter = 0 ;
}

static
void clear_args (struct gengetopt_args_info *args_info)
{
  FIX_UNUSED (args_info);
  args_info->referenceFASTA_arg = NULL;
  args_info->referenceFASTA_orig = NULL;
  args_info->readDepths_arg = gengetopt_strdup ("-");
  args_info->readDepths_orig = NULL;
  args_info->targets_arg = NULL;
  args_info->targets_orig = NULL;
  args_info->mergedTargets_arg = gengetopt_strdup ("-");
  args_info->mergedTargets_orig = NULL;
  args_info->GATKdepths_arg = NULL;
  args_info->GATKdepths_orig = NULL;
  args_info->GATKdepthsList_arg = NULL;
  args_info->GATKdepthsList_orig = NULL;
  args_info->sampleIDmap_arg = NULL;
  args_info->sampleIDmap_orig = NULL;
  args_info->fromID_arg = 1;
  args_info->fromID_orig = NULL;
  args_info->toID_arg = 2;
  args_info->toID_orig = NULL;
  args_info->columnSuffix_arg = gengetopt_strdup ("_mean_cvg");
  args_info->columnSuffix_orig = NULL;
  args_info->rdPrecision_arg = 2;
  args_info->rdPrecision_orig = NULL;
  args_info->outputTargetsBySamples_flag = 0;
  args_info->excludeTargets_arg = NULL;
  args_info->excludeTargets_orig = NULL;
  args_info->excludeChromosomeTargets_arg = NULL;
  args_info->excludeChromosomeTargets_orig = NULL;
  args_info->excludeSamples_arg = NULL;
  args_info->excludeSamples_orig = NULL;
  args_info->minTargetSize_arg = 0;
  args_info->minTargetSize_orig = NULL;
  args_info->maxTargetSize_orig = NULL;
  args_info->minMeanTargetRD_orig = NULL;
  args_info->maxMeanTargetRD_orig = NULL;
  args_info->minSdTargetRD_arg = 0;
  args_info->minSdTargetRD_orig = NULL;
  args_info->maxSdTargetRD_orig = NULL;
  args_info->minMeanSampleRD_orig = NULL;
  args_info->maxMeanSampleRD_orig = NULL;
  args_info->minSdSampleRD_arg = 0;
  args_info->minSdSampleRD_orig = NULL;
  args_info->maxSdSampleRD_orig = NULL;
  args_info->scaleDataBySum_flag = 0;
  args_info->scaleDataBySumType_arg = scaleDataBySumType__NULL;
  args_info->scaleDataBySumType_orig = NULL;
  args_info->scaleDataBySumFactor_arg = 1e6;
  args_info->scaleDataBySumFactor_orig = NULL;
  args_info->log10_orig = NULL;
  args_info->centerData_flag = 0;
  args_info->centerType_arg = centerType__NULL;
  args_info->centerType_orig = NULL;
  args_info->zScoreData_flag = 0;
  args_info->outputExcludedTargets_arg = NULL;
  args_info->outputExcludedTargets_orig = NULL;
  args_info->outputExcludedSamples_arg = NULL;
  args_info->outputExcludedSamples_orig = NULL;
  args_info->outputMatrix_arg = gengetopt_strdup ("-");
  args_info->outputMatrix_orig = NULL;
  args_info->PCA_saveMemory_arg = gengetopt_strdup ("");
  args_info->PCA_saveMemory_orig = NULL;
  args_info->normalizeOutput_arg = gengetopt_strdup ("-");
  args_info->normalizeOutput_orig = NULL;
  args_info->PCnormalizeMethod_arg = PCnormalizeMethod_arg_PVE_mean;
  args_info->PCnormalizeMethod_orig = NULL;
  args_info->numPCtoRemove_arg = 20;
  args_info->numPCtoRemove_orig = NULL;
  args_info->PVE_mean_factor_arg = 0.7;
  args_info->PVE_mean_factor_orig = NULL;
  args_info->PVE_contrib_arg = 50;
  args_info->PVE_contrib_orig = NULL;
  args_info->PCAfiles_arg = NULL;
  args_info->PCAfiles_orig = NULL;
  args_info->xcnv_arg = gengetopt_strdup ("-");
  args_info->xcnv_orig = NULL;
  args_info->discoverSomeQualThresh_arg = 30;
  args_info->discoverSomeQualThresh_orig = NULL;
  args_info->posteriorFiles_arg = NULL;
  args_info->posteriorFiles_orig = NULL;
  args_info->gxcnv_arg = NULL;
  args_info->gxcnv_orig = NULL;
  args_info->subsegments_flag = 0;
  args_info->maxTargetsInSubsegment_arg = 30;
  args_info->maxTargetsInSubsegment_orig = NULL;
  args_info->genotypeQualThresholdWhenNoExact_arg = 20;
  args_info->genotypeQualThresholdWhenNoExact_orig = NULL;
  args_info->mergeVCF_arg = NULL;
  args_info->mergeVCF_orig = NULL;
  args_info->mergeVCFlist_arg = NULL;
  args_info->mergeVCFlist_orig = NULL;
  args_info->paramFile_arg = NULL;
  args_info->paramFile_orig = NULL;
  args_info->DB_arg = NULL;
  args_info->DB_orig = NULL;
  args_info->maxNormalizedReadDepthVal_arg = 10;
  args_info->maxNormalizedReadDepthVal_orig = NULL;
  args_info->maxQualScore_arg = 99;
  args_info->maxQualScore_orig = NULL;
  args_info->scorePrecision_arg = 0;
  args_info->scorePrecision_orig = NULL;
  args_info->aux_xcnv_arg = NULL;
  args_info->aux_xcnv_orig = NULL;
  args_info->auxUpstreamPrintTargs_arg = 2;
  args_info->auxUpstreamPrintTargs_orig = NULL;
  args_info->auxDownstreamPrintTargs_arg = 2;
  args_info->auxDownstreamPrintTargs_orig = NULL;
  args_info->origReadDepths_arg = NULL;
  args_info->origReadDepths_orig = NULL;
  args_info->keepSampleIDs_arg = NULL;
  args_info->keepSampleIDs_orig = NULL;
  args_info->vcf_arg = gengetopt_strdup ("-");
  args_info->vcf_orig = NULL;
  
}

static
void init_args_info(struct gengetopt_args_info *args_info)
{
  init_full_help_array(); 
  init_help_array(); 
  args_info->help_help = gengetopt_args_info_detailed_help[0] ;
  args_info->detailed_help_help = gengetopt_args_info_detailed_help[1] ;
  args_info->full_help_help = gengetopt_args_info_detailed_help[2] ;
  args_info->version_help = gengetopt_args_info_detailed_help[3] ;
  args_info->referenceFASTA_help = gengetopt_args_info_detailed_help[5] ;
  args_info->readDepths_help = gengetopt_args_info_detailed_help[7] ;
  args_info->prepareTargets_help = gengetopt_args_info_detailed_help[11] ;
  args_info->targets_help = gengetopt_args_info_detailed_help[12] ;
  args_info->targets_min = 0;
  args_info->targets_max = 0;
  args_info->mergedTargets_help = gengetopt_args_info_detailed_help[13] ;
  args_info->mergeGATKdepths_help = gengetopt_args_info_detailed_help[15] ;
  args_info->GATKdepths_help = gengetopt_args_info_detailed_help[16] ;
  args_info->GATKdepths_min = 0;
  args_info->GATKdepths_max = 0;
  args_info->GATKdepthsList_help = gengetopt_args_info_detailed_help[17] ;
  args_info->GATKdepthsList_min = 0;
  args_info->GATKdepthsList_max = 0;
  args_info->sampleIDmap_help = gengetopt_args_info_detailed_help[18] ;
  args_info->fromID_help = gengetopt_args_info_detailed_help[19] ;
  args_info->toID_help = gengetopt_args_info_detailed_help[20] ;
  args_info->columnSuffix_help = gengetopt_args_info_detailed_help[21] ;
  args_info->rdPrecision_help = gengetopt_args_info_detailed_help[22] ;
  args_info->outputTargetsBySamples_help = gengetopt_args_info_detailed_help[23] ;
  args_info->matrix_help = gengetopt_args_info_detailed_help[25] ;
  args_info->excludeTargets_help = gengetopt_args_info_detailed_help[26] ;
  args_info->excludeTargets_min = 0;
  args_info->excludeTargets_max = 0;
  args_info->excludeChromosomeTargets_help = gengetopt_args_info_detailed_help[27] ;
  args_info->excludeChromosomeTargets_min = 0;
  args_info->excludeChromosomeTargets_max = 0;
  args_info->excludeSamples_help = gengetopt_args_info_detailed_help[28] ;
  args_info->excludeSamples_min = 0;
  args_info->excludeSamples_max = 0;
  args_info->minTargetSize_help = gengetopt_args_info_detailed_help[29] ;
  args_info->maxTargetSize_help = gengetopt_args_info_detailed_help[30] ;
  args_info->minMeanTargetRD_help = gengetopt_args_info_detailed_help[31] ;
  args_info->maxMeanTargetRD_help = gengetopt_args_info_detailed_help[32] ;
  args_info->minSdTargetRD_help = gengetopt_args_info_detailed_help[33] ;
  args_info->maxSdTargetRD_help = gengetopt_args_info_detailed_help[34] ;
  args_info->minMeanSampleRD_help = gengetopt_args_info_detailed_help[35] ;
  args_info->maxMeanSampleRD_help = gengetopt_args_info_detailed_help[36] ;
  args_info->minSdSampleRD_help = gengetopt_args_info_detailed_help[37] ;
  args_info->maxSdSampleRD_help = gengetopt_args_info_detailed_help[38] ;
  args_info->scaleDataBySum_help = gengetopt_args_info_detailed_help[39] ;
  args_info->scaleDataBySumType_help = gengetopt_args_info_detailed_help[40] ;
  args_info->scaleDataBySumFactor_help = gengetopt_args_info_detailed_help[41] ;
  args_info->log10_help = gengetopt_args_info_detailed_help[42] ;
  args_info->centerData_help = gengetopt_args_info_detailed_help[43] ;
  args_info->centerType_help = gengetopt_args_info_detailed_help[44] ;
  args_info->zScoreData_help = gengetopt_args_info_detailed_help[45] ;
  args_info->outputExcludedTargets_help = gengetopt_args_info_detailed_help[46] ;
  args_info->outputExcludedSamples_help = gengetopt_args_info_detailed_help[47] ;
  args_info->outputMatrix_help = gengetopt_args_info_detailed_help[49] ;
  args_info->PCA_help = gengetopt_args_info_detailed_help[52] ;
  args_info->PCA_saveMemory_help = gengetopt_args_info_detailed_help[53] ;
  args_info->normalize_help = gengetopt_args_info_detailed_help[56] ;
  args_info->normalizeOutput_help = gengetopt_args_info_detailed_help[57] ;
  args_info->PCnormalizeMethod_help = gengetopt_args_info_detailed_help[58] ;
  args_info->numPCtoRemove_help = gengetopt_args_info_detailed_help[59] ;
  args_info->PVE_mean_factor_help = gengetopt_args_info_detailed_help[60] ;
  args_info->PVE_contrib_help = gengetopt_args_info_detailed_help[61] ;
  args_info->PCAfiles_help = gengetopt_args_info_detailed_help[63] ;
  args_info->createDB_help = gengetopt_args_info_detailed_help[65] ;
  args_info->discover_help = gengetopt_args_info_detailed_help[68] ;
  args_info->xcnv_help = gengetopt_args_info_detailed_help[69] ;
  args_info->optDiscover_help = gengetopt_args_info_detailed_help[70] ;
  args_info->discoverSomeQualThresh_help = gengetopt_args_info_detailed_help[71] ;
  args_info->posteriorFiles_help = gengetopt_args_info_detailed_help[72] ;
  args_info->genotype_help = gengetopt_args_info_detailed_help[74] ;
  args_info->gxcnv_help = gengetopt_args_info_detailed_help[75] ;
  args_info->subsegments_help = gengetopt_args_info_detailed_help[76] ;
  args_info->maxTargetsInSubsegment_help = gengetopt_args_info_detailed_help[78] ;
  args_info->genotypeQualThresholdWhenNoExact_help = gengetopt_args_info_detailed_help[79] ;
  args_info->mergeVCFs_help = gengetopt_args_info_detailed_help[82] ;
  args_info->mergeVCF_help = gengetopt_args_info_detailed_help[83] ;
  args_info->mergeVCF_min = 0;
  args_info->mergeVCF_max = 0;
  args_info->mergeVCFlist_help = gengetopt_args_info_detailed_help[84] ;
  args_info->mergeVCFlist_min = 0;
  args_info->mergeVCFlist_max = 0;
  args_info->paramFile_help = gengetopt_args_info_detailed_help[86] ;
  args_info->DB_help = gengetopt_args_info_detailed_help[89] ;
  args_info->maxNormalizedReadDepthVal_help = gengetopt_args_info_detailed_help[91] ;
  args_info->maxQualScore_help = gengetopt_args_info_detailed_help[92] ;
  args_info->scorePrecision_help = gengetopt_args_info_detailed_help[93] ;
  args_info->aux_xcnv_help = gengetopt_args_info_detailed_help[94] ;
  args_info->auxUpstreamPrintTargs_help = gengetopt_args_info_detailed_help[96] ;
  args_info->auxDownstreamPrintTargs_help = gengetopt_args_info_detailed_help[97] ;
  args_info->origReadDepths_help = gengetopt_args_info_detailed_help[98] ;
  args_info->keepSampleIDs_help = gengetopt_args_info_detailed_help[100] ;
  args_info->vcf_help = gengetopt_args_info_detailed_help[103] ;
  args_info->printHMM_help = gengetopt_args_info_detailed_help[106] ;
  args_info->transition_help = gengetopt_args_info_detailed_help[108] ;
  
}

void
cmdline_parser_print_version (void)
{
  printf ("%s %s\n",
     (strlen(CMDLINE_PARSER_PACKAGE_NAME) ? CMDLINE_PARSER_PACKAGE_NAME : CMDLINE_PARSER_PACKAGE),
     CMDLINE_PARSER_VERSION);
}

static void print_help_common(void) {
  cmdline_parser_print_version ();

  if (strlen(gengetopt_args_info_purpose) > 0)
    printf("\n%s\n", gengetopt_args_info_purpose);

  if (strlen(gengetopt_args_info_usage) > 0)
    printf("\n%s\n", gengetopt_args_info_usage);

  printf("\n");

  if (strlen(gengetopt_args_info_description) > 0)
    printf("%s\n\n", gengetopt_args_info_description);
}

void
cmdline_parser_print_help (void)
{
  int i = 0;
  print_help_common();
  while (gengetopt_args_info_help[i])
    printf("%s\n", gengetopt_args_info_help[i++]);
}

void
cmdline_parser_print_full_help (void)
{
  int i = 0;
  print_help_common();
  while (gengetopt_args_info_full_help[i])
    printf("%s\n", gengetopt_args_info_full_help[i++]);
}

void
cmdline_parser_print_detailed_help (void)
{
  int i = 0;
  print_help_common();
  while (gengetopt_args_info_detailed_help[i])
    printf("%s\n", gengetopt_args_info_detailed_help[i++]);
}

void
cmdline_parser_init (struct gengetopt_args_info *args_info)
{
  clear_given (args_info);
  clear_args (args_info);
  init_args_info (args_info);
}

void
cmdline_parser_params_init(struct cmdline_parser_params *params)
{
  if (params)
    { 
      params->override = 0;
      params->initialize = 1;
      params->check_required = 1;
      params->check_ambiguity = 0;
      params->print_errors = 1;
    }
}

struct cmdline_parser_params *
cmdline_parser_params_create(void)
{
  struct cmdline_parser_params *params = 
    (struct cmdline_parser_params *)malloc(sizeof(struct cmdline_parser_params));
  cmdline_parser_params_init(params);  
  return params;
}

static void
free_string_field (char **s)
{
  if (*s)
    {
      free (*s);
      *s = 0;
    }
}

/** @brief generic value variable */
union generic_value {
    int int_arg;
    double double_arg;
    char *string_arg;
    const char *default_string_arg;
};

/** @brief holds temporary values for multiple options */
struct generic_list
{
  union generic_value arg;
  char *orig;
  struct generic_list *next;
};

/**
 * @brief add a node at the head of the list 
 */
static void add_node(struct generic_list **list) {
  struct generic_list *new_node = (struct generic_list *) malloc (sizeof (struct generic_list));
  new_node->next = *list;
  *list = new_node;
  new_node->arg.string_arg = 0;
  new_node->orig = 0;
}


static void
free_multiple_string_field(unsigned int len, char ***arg, char ***orig)
{
  unsigned int i;
  if (*arg) {
    for (i = 0; i < len; ++i)
      {
        free_string_field(&((*arg)[i]));
        free_string_field(&((*orig)[i]));
      }
    free_string_field(&((*arg)[0])); /* free default string */

    free (*arg);
    *arg = 0;
    free (*orig);
    *orig = 0;
  }
}

static void
cmdline_parser_release (struct gengetopt_args_info *args_info)
{

  free_string_field (&(args_info->referenceFASTA_arg));
  free_string_field (&(args_info->referenceFASTA_orig));
  free_string_field (&(args_info->readDepths_arg));
  free_string_field (&(args_info->readDepths_orig));
  free_multiple_string_field (args_info->targets_given, &(args_info->targets_arg), &(args_info->targets_orig));
  free_string_field (&(args_info->mergedTargets_arg));
  free_string_field (&(args_info->mergedTargets_orig));
  free_multiple_string_field (args_info->GATKdepths_given, &(args_info->GATKdepths_arg), &(args_info->GATKdepths_orig));
  free_multiple_string_field (args_info->GATKdepthsList_given, &(args_info->GATKdepthsList_arg), &(args_info->GATKdepthsList_orig));
  free_string_field (&(args_info->sampleIDmap_arg));
  free_string_field (&(args_info->sampleIDmap_orig));
  free_string_field (&(args_info->fromID_orig));
  free_string_field (&(args_info->toID_orig));
  free_string_field (&(args_info->columnSuffix_arg));
  free_string_field (&(args_info->columnSuffix_orig));
  free_string_field (&(args_info->rdPrecision_orig));
  free_multiple_string_field (args_info->excludeTargets_given, &(args_info->excludeTargets_arg), &(args_info->excludeTargets_orig));
  free_multiple_string_field (args_info->excludeChromosomeTargets_given, &(args_info->excludeChromosomeTargets_arg), &(args_info->excludeChromosomeTargets_orig));
  free_multiple_string_field (args_info->excludeSamples_given, &(args_info->excludeSamples_arg), &(args_info->excludeSamples_orig));
  free_string_field (&(args_info->minTargetSize_orig));
  free_string_field (&(args_info->maxTargetSize_orig));
  free_string_field (&(args_info->minMeanTargetRD_orig));
  free_string_field (&(args_info->maxMeanTargetRD_orig));
  free_string_field (&(args_info->minSdTargetRD_orig));
  free_string_field (&(args_info->maxSdTargetRD_orig));
  free_string_field (&(args_info->minMeanSampleRD_orig));
  free_string_field (&(args_info->maxMeanSampleRD_orig));
  free_string_field (&(args_info->minSdSampleRD_orig));
  free_string_field (&(args_info->maxSdSampleRD_orig));
  free_string_field (&(args_info->scaleDataBySumType_orig));
  free_string_field (&(args_info->scaleDataBySumFactor_orig));
  free_string_field (&(args_info->log10_orig));
  free_string_field (&(args_info->centerType_orig));
  free_string_field (&(args_info->outputExcludedTargets_arg));
  free_string_field (&(args_info->outputExcludedTargets_orig));
  free_string_field (&(args_info->outputExcludedSamples_arg));
  free_string_field (&(args_info->outputExcludedSamples_orig));
  free_string_field (&(args_info->outputMatrix_arg));
  free_string_field (&(args_info->outputMatrix_orig));
  free_string_field (&(args_info->PCA_saveMemory_arg));
  free_string_field (&(args_info->PCA_saveMemory_orig));
  free_string_field (&(args_info->normalizeOutput_arg));
  free_string_field (&(args_info->normalizeOutput_orig));
  free_string_field (&(args_info->PCnormalizeMethod_orig));
  free_string_field (&(args_info->numPCtoRemove_orig));
  free_string_field (&(args_info->PVE_mean_factor_orig));
  free_string_field (&(args_info->PVE_contrib_orig));
  free_string_field (&(args_info->PCAfiles_arg));
  free_string_field (&(args_info->PCAfiles_orig));
  free_string_field (&(args_info->xcnv_arg));
  free_string_field (&(args_info->xcnv_orig));
  free_string_field (&(args_info->discoverSomeQualThresh_orig));
  free_string_field (&(args_info->posteriorFiles_arg));
  free_string_field (&(args_info->posteriorFiles_orig));
  free_string_field (&(args_info->gxcnv_arg));
  free_string_field (&(args_info->gxcnv_orig));
  free_string_field (&(args_info->maxTargetsInSubsegment_orig));
  free_string_field (&(args_info->genotypeQualThresholdWhenNoExact_orig));
  free_multiple_string_field (args_info->mergeVCF_given, &(args_info->mergeVCF_arg), &(args_info->mergeVCF_orig));
  free_multiple_string_field (args_info->mergeVCFlist_given, &(args_info->mergeVCFlist_arg), &(args_info->mergeVCFlist_orig));
  free_string_field (&(args_info->paramFile_arg));
  free_string_field (&(args_info->paramFile_orig));
  free_string_field (&(args_info->DB_arg));
  free_string_field (&(args_info->DB_orig));
  free_string_field (&(args_info->maxNormalizedReadDepthVal_orig));
  free_string_field (&(args_info->maxQualScore_orig));
  free_string_field (&(args_info->scorePrecision_orig));
  free_string_field (&(args_info->aux_xcnv_arg));
  free_string_field (&(args_info->aux_xcnv_orig));
  free_string_field (&(args_info->auxUpstreamPrintTargs_orig));
  free_string_field (&(args_info->auxDownstreamPrintTargs_orig));
  free_string_field (&(args_info->origReadDepths_arg));
  free_string_field (&(args_info->origReadDepths_orig));
  free_string_field (&(args_info->keepSampleIDs_arg));
  free_string_field (&(args_info->keepSampleIDs_orig));
  free_string_field (&(args_info->vcf_arg));
  free_string_field (&(args_info->vcf_orig));
  
  

  clear_given (args_info);
}

/**
 * @param val the value to check
 * @param values the possible values
 * @return the index of the matched value:
 * -1 if no value matched,
 * -2 if more than one value has matched
 */
static int
check_possible_values(const char *val, const char *values[])
{
  int i, found, last;
  size_t len;

  if (!val)   /* otherwise strlen() crashes below */
    return -1; /* -1 means no argument for the option */

  found = last = 0;

  for (i = 0, len = strlen(val); values[i]; ++i)
    {
      if (strncmp(val, values[i], len) == 0)
        {
          ++found;
          last = i;
          if (strlen(values[i]) == len)
            return i; /* exact macth no need to check more */
        }
    }

  if (found == 1) /* one match: OK */
    return last;

  return (found ? -2 : -1); /* return many values or none matched */
}


static void
write_into_file(FILE *outfile, const char *opt, const char *arg, const char *values[])
{
  int found = -1;
  if (arg) {
    if (values) {
      found = check_possible_values(arg, values);      
    }
    if (found >= 0)
      fprintf(outfile, "%s=\"%s\" # %s\n", opt, arg, values[found]);
    else
      fprintf(outfile, "%s=\"%s\"\n", opt, arg);
  } else {
    fprintf(outfile, "%s\n", opt);
  }
}

static void
write_multiple_into_file(FILE *outfile, int len, const char *opt, char **arg, const char *values[])
{
  int i;
  
  for (i = 0; i < len; ++i)
    write_into_file(outfile, opt, (arg ? arg[i] : 0), values);
}

int
cmdline_parser_dump(FILE *outfile, struct gengetopt_args_info *args_info)
{
  int i = 0;

  if (!outfile)
    {
      fprintf (stderr, "%s: cannot dump options to stream\n", CMDLINE_PARSER_PACKAGE);
      return EXIT_FAILURE;
    }

  if (args_info->help_given)
    write_into_file(outfile, "help", 0, 0 );
  if (args_info->detailed_help_given)
    write_into_file(outfile, "detailed-help", 0, 0 );
  if (args_info->full_help_given)
    write_into_file(outfile, "full-help", 0, 0 );
  if (args_info->version_given)
    write_into_file(outfile, "version", 0, 0 );
  if (args_info->referenceFASTA_given)
    write_into_file(outfile, "referenceFASTA", args_info->referenceFASTA_orig, 0);
  if (args_info->readDepths_given)
    write_into_file(outfile, "readDepths", args_info->readDepths_orig, 0);
  if (args_info->prepareTargets_given)
    write_into_file(outfile, "prepareTargets", 0, 0 );
  write_multiple_into_file(outfile, args_info->targets_given, "targets", args_info->targets_orig, 0);
  if (args_info->mergedTargets_given)
    write_into_file(outfile, "mergedTargets", args_info->mergedTargets_orig, 0);
  if (args_info->mergeGATKdepths_given)
    write_into_file(outfile, "mergeGATKdepths", 0, 0 );
  write_multiple_into_file(outfile, args_info->GATKdepths_given, "GATKdepths", args_info->GATKdepths_orig, 0);
  write_multiple_into_file(outfile, args_info->GATKdepthsList_given, "GATKdepthsList", args_info->GATKdepthsList_orig, 0);
  if (args_info->sampleIDmap_given)
    write_into_file(outfile, "sampleIDmap", args_info->sampleIDmap_orig, 0);
  if (args_info->fromID_given)
    write_into_file(outfile, "fromID", args_info->fromID_orig, 0);
  if (args_info->toID_given)
    write_into_file(outfile, "toID", args_info->toID_orig, 0);
  if (args_info->columnSuffix_given)
    write_into_file(outfile, "columnSuffix", args_info->columnSuffix_orig, 0);
  if (args_info->rdPrecision_given)
    write_into_file(outfile, "rdPrecision", args_info->rdPrecision_orig, 0);
  if (args_info->outputTargetsBySamples_given)
    write_into_file(outfile, "outputTargetsBySamples", 0, 0 );
  if (args_info->matrix_given)
    write_into_file(outfile, "matrix", 0, 0 );
  write_multiple_into_file(outfile, args_info->excludeTargets_given, "excludeTargets", args_info->excludeTargets_orig, 0);
  write_multiple_into_file(outfile, args_info->excludeChromosomeTargets_given, "excludeChromosomeTargets", args_info->excludeChromosomeTargets_orig, 0);
  write_multiple_into_file(outfile, args_info->excludeSamples_given, "excludeSamples", args_info->excludeSamples_orig, 0);
  if (args_info->minTargetSize_given)
    write_into_file(outfile, "minTargetSize", args_info->minTargetSize_orig, 0);
  if (args_info->maxTargetSize_given)
    write_into_file(outfile, "maxTargetSize", args_info->maxTargetSize_orig, 0);
  if (args_info->minMeanTargetRD_given)
    write_into_file(outfile, "minMeanTargetRD", args_info->minMeanTargetRD_orig, 0);
  if (args_info->maxMeanTargetRD_given)
    write_into_file(outfile, "maxMeanTargetRD", args_info->maxMeanTargetRD_orig, 0);
  if (args_info->minSdTargetRD_given)
    write_into_file(outfile, "minSdTargetRD", args_info->minSdTargetRD_orig, 0);
  if (args_info->maxSdTargetRD_given)
    write_into_file(outfile, "maxSdTargetRD", args_info->maxSdTargetRD_orig, 0);
  if (args_info->minMeanSampleRD_given)
    write_into_file(outfile, "minMeanSampleRD", args_info->minMeanSampleRD_orig, 0);
  if (args_info->maxMeanSampleRD_given)
    write_into_file(outfile, "maxMeanSampleRD", args_info->maxMeanSampleRD_orig, 0);
  if (args_info->minSdSampleRD_given)
    write_into_file(outfile, "minSdSampleRD", args_info->minSdSampleRD_orig, 0);
  if (args_info->maxSdSampleRD_given)
    write_into_file(outfile, "maxSdSampleRD", args_info->maxSdSampleRD_orig, 0);
  if (args_info->scaleDataBySum_given)
    write_into_file(outfile, "scaleDataBySum", 0, 0 );
  if (args_info->scaleDataBySumType_given)
    write_into_file(outfile, "scaleDataBySumType", args_info->scaleDataBySumType_orig, cmdline_parser_scaleDataBySumType_values);
  if (args_info->scaleDataBySumFactor_given)
    write_into_file(outfile, "scaleDataBySumFactor", args_info->scaleDataBySumFactor_orig, 0);
  if (args_info->log10_given)
    write_into_file(outfile, "log10", args_info->log10_orig, 0);
  if (args_info->centerData_given)
    write_into_file(outfile, "centerData", 0, 0 );
  if (args_info->centerType_given)
    write_into_file(outfile, "centerType", args_info->centerType_orig, cmdline_parser_centerType_values);
  if (args_info->zScoreData_given)
    write_into_file(outfile, "zScoreData", 0, 0 );
  if (args_info->outputExcludedTargets_given)
    write_into_file(outfile, "outputExcludedTargets", args_info->outputExcludedTargets_orig, 0);
  if (args_info->outputExcludedSamples_given)
    write_into_file(outfile, "outputExcludedSamples", args_info->outputExcludedSamples_orig, 0);
  if (args_info->outputMatrix_given)
    write_into_file(outfile, "outputMatrix", args_info->outputMatrix_orig, 0);
  if (args_info->PCA_given)
    write_into_file(outfile, "PCA", 0, 0 );
  if (args_info->PCA_saveMemory_given)
    write_into_file(outfile, "PCA_saveMemory", args_info->PCA_saveMemory_orig, 0);
  if (args_info->normalize_given)
    write_into_file(outfile, "normalize", 0, 0 );
  if (args_info->normalizeOutput_given)
    write_into_file(outfile, "normalizeOutput", args_info->normalizeOutput_orig, 0);
  if (args_info->PCnormalizeMethod_given)
    write_into_file(outfile, "PCnormalizeMethod", args_info->PCnormalizeMethod_orig, cmdline_parser_PCnormalizeMethod_values);
  if (args_info->numPCtoRemove_given)
    write_into_file(outfile, "numPCtoRemove", args_info->numPCtoRemove_orig, 0);
  if (args_info->PVE_mean_factor_given)
    write_into_file(outfile, "PVE_mean_factor", args_info->PVE_mean_factor_orig, 0);
  if (args_info->PVE_contrib_given)
    write_into_file(outfile, "PVE_contrib", args_info->PVE_contrib_orig, 0);
  if (args_info->PCAfiles_given)
    write_into_file(outfile, "PCAfiles", args_info->PCAfiles_orig, 0);
  if (args_info->createDB_given)
    write_into_file(outfile, "createDB", 0, 0 );
  if (args_info->discover_given)
    write_into_file(outfile, "discover", 0, 0 );
  if (args_info->xcnv_given)
    write_into_file(outfile, "xcnv", args_info->xcnv_orig, 0);
  if (args_info->optDiscover_given)
    write_into_file(outfile, "optDiscover", 0, 0 );
  if (args_info->discoverSomeQualThresh_given)
    write_into_file(outfile, "discoverSomeQualThresh", args_info->discoverSomeQualThresh_orig, 0);
  if (args_info->posteriorFiles_given)
    write_into_file(outfile, "posteriorFiles", args_info->posteriorFiles_orig, 0);
  if (args_info->genotype_given)
    write_into_file(outfile, "genotype", 0, 0 );
  if (args_info->gxcnv_given)
    write_into_file(outfile, "gxcnv", args_info->gxcnv_orig, 0);
  if (args_info->subsegments_given)
    write_into_file(outfile, "subsegments", 0, 0 );
  if (args_info->maxTargetsInSubsegment_given)
    write_into_file(outfile, "maxTargetsInSubsegment", args_info->maxTargetsInSubsegment_orig, 0);
  if (args_info->genotypeQualThresholdWhenNoExact_given)
    write_into_file(outfile, "genotypeQualThresholdWhenNoExact", args_info->genotypeQualThresholdWhenNoExact_orig, 0);
  if (args_info->mergeVCFs_given)
    write_into_file(outfile, "mergeVCFs", 0, 0 );
  write_multiple_into_file(outfile, args_info->mergeVCF_given, "mergeVCF", args_info->mergeVCF_orig, 0);
  write_multiple_into_file(outfile, args_info->mergeVCFlist_given, "mergeVCFlist", args_info->mergeVCFlist_orig, 0);
  if (args_info->paramFile_given)
    write_into_file(outfile, "paramFile", args_info->paramFile_orig, 0);
  if (args_info->DB_given)
    write_into_file(outfile, "DB", args_info->DB_orig, 0);
  if (args_info->maxNormalizedReadDepthVal_given)
    write_into_file(outfile, "maxNormalizedReadDepthVal", args_info->maxNormalizedReadDepthVal_orig, 0);
  if (args_info->maxQualScore_given)
    write_into_file(outfile, "maxQualScore", args_info->maxQualScore_orig, 0);
  if (args_info->scorePrecision_given)
    write_into_file(outfile, "scorePrecision", args_info->scorePrecision_orig, 0);
  if (args_info->aux_xcnv_given)
    write_into_file(outfile, "aux_xcnv", args_info->aux_xcnv_orig, 0);
  if (args_info->auxUpstreamPrintTargs_given)
    write_into_file(outfile, "auxUpstreamPrintTargs", args_info->auxUpstreamPrintTargs_orig, 0);
  if (args_info->auxDownstreamPrintTargs_given)
    write_into_file(outfile, "auxDownstreamPrintTargs", args_info->auxDownstreamPrintTargs_orig, 0);
  if (args_info->origReadDepths_given)
    write_into_file(outfile, "origReadDepths", args_info->origReadDepths_orig, 0);
  if (args_info->keepSampleIDs_given)
    write_into_file(outfile, "keepSampleIDs", args_info->keepSampleIDs_orig, 0);
  if (args_info->vcf_given)
    write_into_file(outfile, "vcf", args_info->vcf_orig, 0);
  if (args_info->printHMM_given)
    write_into_file(outfile, "printHMM", 0, 0 );
  if (args_info->transition_given)
    write_into_file(outfile, "transition", 0, 0 );
  

  i = EXIT_SUCCESS;
  return i;
}

int
cmdline_parser_file_save(const char *filename, struct gengetopt_args_info *args_info)
{
  FILE *outfile;
  int i = 0;

  outfile = fopen(filename, "w");

  if (!outfile)
    {
      fprintf (stderr, "%s: cannot open file for writing: %s\n", CMDLINE_PARSER_PACKAGE, filename);
      return EXIT_FAILURE;
    }

  i = cmdline_parser_dump(outfile, args_info);
  fclose (outfile);

  return i;
}

void
cmdline_parser_free (struct gengetopt_args_info *args_info)
{
  cmdline_parser_release (args_info);
}

/** @brief replacement of strdup, which is not standard */
char *
gengetopt_strdup (const char *s)
{
  char *result = 0;
  if (!s)
    return result;

  result = (char*)malloc(strlen(s) + 1);
  if (result == (char*)0)
    return (char*)0;
  strcpy(result, s);
  return result;
}

static char *
get_multiple_arg_token(const char *arg)
{
  const char *tok;
  char *ret;
  size_t len, num_of_escape, i, j;

  if (!arg)
    return 0;

  tok = strchr (arg, ',');
  num_of_escape = 0;

  /* make sure it is not escaped */
  while (tok)
    {
      if (*(tok-1) == '\\')
        {
          /* find the next one */
          tok = strchr (tok+1, ',');
          ++num_of_escape;
        }
      else
        break;
    }

  if (tok)
    len = (size_t)(tok - arg + 1);
  else
    len = strlen (arg) + 1;

  len -= num_of_escape;

  ret = (char *) malloc (len);

  i = 0;
  j = 0;
  while (arg[i] && (j < len-1))
    {
      if (arg[i] == '\\' && 
	  arg[ i + 1 ] && 
	  arg[ i + 1 ] == ',')
        ++i;

      ret[j++] = arg[i++];
    }

  ret[len-1] = '\0';

  return ret;
}

static const char *
get_multiple_arg_token_next(const char *arg)
{
  const char *tok;

  if (!arg)
    return 0;

  tok = strchr (arg, ',');

  /* make sure it is not escaped */
  while (tok)
    {
      if (*(tok-1) == '\\')
        {
          /* find the next one */
          tok = strchr (tok+1, ',');
        }
      else
        break;
    }

  if (! tok || strlen(tok) == 1)
    return 0;

  return tok+1;
}

static int
check_multiple_option_occurrences(const char *prog_name, unsigned int option_given, unsigned int min, unsigned int max, const char *option_desc);

int
check_multiple_option_occurrences(const char *prog_name, unsigned int option_given, unsigned int min, unsigned int max, const char *option_desc)
{
  int error = 0;

  if (option_given && (min > 0 || max > 0))
    {
      if (min > 0 && max > 0)
        {
          if (min == max)
            {
              /* specific occurrences */
              if (option_given != (unsigned int) min)
                {
                  fprintf (stderr, "%s: %s option occurrences must be %d\n",
                    prog_name, option_desc, min);
                  error = 1;
                }
            }
          else if (option_given < (unsigned int) min
                || option_given > (unsigned int) max)
            {
              /* range occurrences */
              fprintf (stderr, "%s: %s option occurrences must be between %d and %d\n",
                prog_name, option_desc, min, max);
              error = 1;
            }
        }
      else if (min > 0)
        {
          /* at least check */
          if (option_given < min)
            {
              fprintf (stderr, "%s: %s option occurrences must be at least %d\n",
                prog_name, option_desc, min);
              error = 1;
            }
        }
      else if (max > 0)
        {
          /* at most check */
          if (option_given > max)
            {
              fprintf (stderr, "%s: %s option occurrences must be at most %d\n",
                prog_name, option_desc, max);
              error = 1;
            }
        }
    }
    
  return error;
}
int
cmdline_parser (int argc, char **argv, struct gengetopt_args_info *args_info)
{
  return cmdline_parser2 (argc, argv, args_info, 0, 1, 1);
}

int
cmdline_parser_ext (int argc, char **argv, struct gengetopt_args_info *args_info,
                   struct cmdline_parser_params *params)
{
  int result;
  result = cmdline_parser_internal (argc, argv, args_info, params, 0);

  if (result == EXIT_FAILURE)
    {
      cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}

int
cmdline_parser2 (int argc, char **argv, struct gengetopt_args_info *args_info, int override, int initialize, int check_required)
{
  int result;
  struct cmdline_parser_params params;
  
  params.override = override;
  params.initialize = initialize;
  params.check_required = check_required;
  params.check_ambiguity = 0;
  params.print_errors = 1;

  result = cmdline_parser_internal (argc, argv, args_info, &params, 0);

  if (result == EXIT_FAILURE)
    {
      cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}

int
cmdline_parser_required (struct gengetopt_args_info *args_info, const char *prog_name)
{
  int result = EXIT_SUCCESS;

  if (cmdline_parser_required2(args_info, prog_name, 0) > 0)
    result = EXIT_FAILURE;

  if (result == EXIT_FAILURE)
    {
      cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}

int
cmdline_parser_required2 (struct gengetopt_args_info *args_info, const char *prog_name, const char *additional_error)
{
  int error = 0;
  FIX_UNUSED (additional_error);

  /* checks for required options */
  if (args_info->prepareTargets_mode_counter && ! args_info->prepareTargets_given)
    {
      fprintf (stderr, "%s: '--prepareTargets' ('-S') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->prepareTargets_mode_counter && ! args_info->targets_given)
    {
      fprintf (stderr, "%s: '--targets' option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->prepareTargets_mode_counter && check_multiple_option_occurrences(prog_name, args_info->targets_given, args_info->targets_min, args_info->targets_max, "'--targets'"))
     error = 1;
  
  if (args_info->mergeGATKdepths_mode_counter && ! args_info->mergeGATKdepths_given)
    {
      fprintf (stderr, "%s: '--mergeGATKdepths' ('-A') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->mergeGATKdepths_mode_counter && check_multiple_option_occurrences(prog_name, args_info->GATKdepths_given, args_info->GATKdepths_min, args_info->GATKdepths_max, "'--GATKdepths'"))
     error = 1;
  
  if (args_info->mergeGATKdepths_mode_counter && check_multiple_option_occurrences(prog_name, args_info->GATKdepthsList_given, args_info->GATKdepthsList_min, args_info->GATKdepthsList_max, "'--GATKdepthsList'"))
     error = 1;
  
  if (args_info->matrix_mode_counter && ! args_info->matrix_given)
    {
      fprintf (stderr, "%s: '--matrix' ('-M') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->matrix_mode_counter && check_multiple_option_occurrences(prog_name, args_info->excludeTargets_given, args_info->excludeTargets_min, args_info->excludeTargets_max, "'--excludeTargets'"))
     error = 1;
  
  if (args_info->matrix_mode_counter && check_multiple_option_occurrences(prog_name, args_info->excludeChromosomeTargets_given, args_info->excludeChromosomeTargets_min, args_info->excludeChromosomeTargets_max, "'--excludeChromosomeTargets'"))
     error = 1;
  
  if (args_info->matrix_mode_counter && check_multiple_option_occurrences(prog_name, args_info->excludeSamples_given, args_info->excludeSamples_min, args_info->excludeSamples_max, "'--excludeSamples'"))
     error = 1;
  
  if (args_info->PCA_mode_counter && ! args_info->PCA_given)
    {
      fprintf (stderr, "%s: '--PCA' ('-P') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->normalize_mode_counter && ! args_info->normalize_given)
    {
      fprintf (stderr, "%s: '--normalize' ('-N') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->discover_mode_counter && ! args_info->discover_given)
    {
      fprintf (stderr, "%s: '--discover' ('-D') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->genotype_mode_counter && ! args_info->genotype_given)
    {
      fprintf (stderr, "%s: '--genotype' ('-G') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->genotype_mode_counter && ! args_info->gxcnv_given)
    {
      fprintf (stderr, "%s: '--gxcnv' ('-g') option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->mergeVCFs_mode_counter && ! args_info->mergeVCFs_given)
    {
      fprintf (stderr, "%s: '--mergeVCFs' option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->mergeVCFs_mode_counter && check_multiple_option_occurrences(prog_name, args_info->mergeVCF_given, args_info->mergeVCF_min, args_info->mergeVCF_max, "'--mergeVCF'"))
     error = 1;
  
  if (args_info->mergeVCFs_mode_counter && check_multiple_option_occurrences(prog_name, args_info->mergeVCFlist_given, args_info->mergeVCFlist_min, args_info->mergeVCFlist_max, "'--mergeVCFlist'"))
     error = 1;
  
  if (args_info->printHMM_mode_counter && ! args_info->printHMM_given)
    {
      fprintf (stderr, "%s: '--printHMM' option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  if (args_info->transition_mode_counter && ! args_info->transition_given)
    {
      fprintf (stderr, "%s: '--transition' option required%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  
  
  /* checks for dependences among options */
  if (args_info->scaleDataBySum_given && ! args_info->scaleDataBySumType_given)
    {
      fprintf (stderr, "%s: '--scaleDataBySum' option depends on option 'scaleDataBySumType'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->scaleDataBySumFactor_given && ! args_info->scaleDataBySumType_given)
    {
      fprintf (stderr, "%s: '--scaleDataBySumFactor' option depends on option 'scaleDataBySumType'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->centerData_given && ! args_info->centerType_given)
    {
      fprintf (stderr, "%s: '--centerData' option depends on option 'centerType'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->zScoreData_given && ! args_info->centerData_given)
    {
      fprintf (stderr, "%s: '--zScoreData' option depends on option 'centerData'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->PCA_given && ! args_info->PCAfiles_given)
    {
      fprintf (stderr, "%s: '--PCA' ('-P') option depends on option 'PCAfiles'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->normalize_given && ! args_info->PCAfiles_given)
    {
      fprintf (stderr, "%s: '--normalize' ('-N') option depends on option 'PCAfiles'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->createDB_given && ! args_info->paramFile_given)
    {
      fprintf (stderr, "%s: '--createDB' ('-C') option depends on option 'paramFile'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->discover_given && ! args_info->paramFile_given)
    {
      fprintf (stderr, "%s: '--discover' ('-D') option depends on option 'paramFile'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->genotype_given && ! args_info->paramFile_given)
    {
      fprintf (stderr, "%s: '--genotype' ('-G') option depends on option 'paramFile'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->maxTargetsInSubsegment_given && ! args_info->subsegments_given)
    {
      fprintf (stderr, "%s: '--maxTargetsInSubsegment' option depends on option 'subsegments'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->auxUpstreamPrintTargs_given && ! args_info->aux_xcnv_given)
    {
      fprintf (stderr, "%s: '--auxUpstreamPrintTargs' ('-u') option depends on option 'aux_xcnv'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->auxDownstreamPrintTargs_given && ! args_info->aux_xcnv_given)
    {
      fprintf (stderr, "%s: '--auxDownstreamPrintTargs' ('-w') option depends on option 'aux_xcnv'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->printHMM_given && ! args_info->paramFile_given)
    {
      fprintf (stderr, "%s: '--printHMM' option depends on option 'paramFile'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }
  if (args_info->transition_given && ! args_info->paramFile_given)
    {
      fprintf (stderr, "%s: '--transition' option depends on option 'paramFile'%s\n", prog_name, (additional_error ? additional_error : ""));
      error = 1;
    }

  return error;
}


static char *package_name = 0;

/**
 * @brief updates an option
 * @param field the generic pointer to the field to update
 * @param orig_field the pointer to the orig field
 * @param field_given the pointer to the number of occurrence of this option
 * @param prev_given the pointer to the number of occurrence already seen
 * @param value the argument for this option (if null no arg was specified)
 * @param possible_values the possible values for this option (if specified)
 * @param default_value the default value (in case the option only accepts fixed values)
 * @param arg_type the type of this option
 * @param check_ambiguity @see cmdline_parser_params.check_ambiguity
 * @param override @see cmdline_parser_params.override
 * @param no_free whether to free a possible previous value
 * @param multiple_option whether this is a multiple option
 * @param long_opt the corresponding long option
 * @param short_opt the corresponding short option (or '-' if none)
 * @param additional_error possible further error specification
 */
static
int update_arg(void *field, char **orig_field,
               unsigned int *field_given, unsigned int *prev_given, 
               char *value, const char *possible_values[],
               const char *default_value,
               cmdline_parser_arg_type arg_type,
               int check_ambiguity, int override,
               int no_free, int multiple_option,
               const char *long_opt, char short_opt,
               const char *additional_error)
{
  char *stop_char = 0;
  const char *val = value;
  int found;
  char **string_field;
  FIX_UNUSED (field);

  stop_char = 0;
  found = 0;

  if (!multiple_option && prev_given && (*prev_given || (check_ambiguity && *field_given)))
    {
      if (short_opt != '-')
        fprintf (stderr, "%s: `--%s' (`-%c') option given more than once%s\n", 
               package_name, long_opt, short_opt,
               (additional_error ? additional_error : ""));
      else
        fprintf (stderr, "%s: `--%s' option given more than once%s\n", 
               package_name, long_opt,
               (additional_error ? additional_error : ""));
      return 1; /* failure */
    }

  if (possible_values && (found = check_possible_values((value ? value : default_value), possible_values)) < 0)
    {
      if (short_opt != '-')
        fprintf (stderr, "%s: %s argument, \"%s\", for option `--%s' (`-%c')%s\n", 
          package_name, (found == -2) ? "ambiguous" : "invalid", value, long_opt, short_opt,
          (additional_error ? additional_error : ""));
      else
        fprintf (stderr, "%s: %s argument, \"%s\", for option `--%s'%s\n", 
          package_name, (found == -2) ? "ambiguous" : "invalid", value, long_opt,
          (additional_error ? additional_error : ""));
      return 1; /* failure */
    }
    
  if (field_given && *field_given && ! override)
    return 0;
  if (prev_given)
    (*prev_given)++;
  if (field_given)
    (*field_given)++;
  if (possible_values)
    val = possible_values[found];

  switch(arg_type) {
  case ARG_FLAG:
    *((int *)field) = !*((int *)field);
    break;
  case ARG_INT:
    if (val) *((int *)field) = strtol (val, &stop_char, 0);
    break;
  case ARG_DOUBLE:
    if (val) *((double *)field) = strtod (val, &stop_char);
    break;
  case ARG_ENUM:
    if (val) *((int *)field) = found;
    break;
  case ARG_STRING:
    if (val) {
      string_field = (char **)field;
      if (!no_free && *string_field)
        free (*string_field); /* free previous string */
      *string_field = gengetopt_strdup (val);
    }
    break;
  default:
    break;
  };

  /* check numeric conversion */
  switch(arg_type) {
  case ARG_INT:
  case ARG_DOUBLE:
    if (val && !(stop_char && *stop_char == '\0')) {
      fprintf(stderr, "%s: invalid numeric value: %s\n", package_name, val);
      return 1; /* failure */
    }
    break;
  default:
    ;
  };

  /* store the original value */
  switch(arg_type) {
  case ARG_NO:
  case ARG_FLAG:
    break;
  default:
    if (value && orig_field) {
      if (no_free) {
        *orig_field = value;
      } else {
        if (*orig_field)
          free (*orig_field); /* free previous string */
        *orig_field = gengetopt_strdup (value);
      }
    }
  };

  return 0; /* OK */
}

/**
 * @brief store information about a multiple option in a temporary list
 * @param list where to (temporarily) store multiple options
 */
static
int update_multiple_arg_temp(struct generic_list **list,
               unsigned int *prev_given, const char *val,
               const char *possible_values[], const char *default_value,
               cmdline_parser_arg_type arg_type,
               const char *long_opt, char short_opt,
               const char *additional_error)
{
  /* store single arguments */
  char *multi_token;
  const char *multi_next;

  if (arg_type == ARG_NO) {
    (*prev_given)++;
    return 0; /* OK */
  }

  multi_token = get_multiple_arg_token(val);
  multi_next = get_multiple_arg_token_next (val);

  while (1)
    {
      add_node (list);
      if (update_arg((void *)&((*list)->arg), &((*list)->orig), 0,
          prev_given, multi_token, possible_values, default_value, 
          arg_type, 0, 1, 1, 1, long_opt, short_opt, additional_error)) {
        if (multi_token) free(multi_token);
        return 1; /* failure */
      }

      if (multi_next)
        {
          multi_token = get_multiple_arg_token(multi_next);
          multi_next = get_multiple_arg_token_next (multi_next);
        }
      else
        break;
    }

  return 0; /* OK */
}

/**
 * @brief free the passed list (including possible string argument)
 */
static
void free_list(struct generic_list *list, short string_arg)
{
  if (list) {
    struct generic_list *tmp;
    while (list)
      {
        tmp = list;
        if (string_arg && list->arg.string_arg)
          free (list->arg.string_arg);
        if (list->orig)
          free (list->orig);
        list = list->next;
        free (tmp);
      }
  }
}

/**
 * @brief updates a multiple option starting from the passed list
 */
static
void update_multiple_arg(void *field, char ***orig_field,
               unsigned int field_given, unsigned int prev_given, union generic_value *default_value,
               cmdline_parser_arg_type arg_type,
               struct generic_list *list)
{
  int i;
  struct generic_list *tmp;

  if (prev_given && list) {
    *orig_field = (char **) realloc (*orig_field, (field_given + prev_given) * sizeof (char *));

    switch(arg_type) {
    case ARG_INT:
    case ARG_ENUM:
      *((int **)field) = (int *)realloc (*((int **)field), (field_given + prev_given) * sizeof (int)); break;
    case ARG_DOUBLE:
      *((double **)field) = (double *)realloc (*((double **)field), (field_given + prev_given) * sizeof (double)); break;
    case ARG_STRING:
      *((char ***)field) = (char **)realloc (*((char ***)field), (field_given + prev_given) * sizeof (char *)); break;
    default:
      break;
    };
    
    for (i = (prev_given - 1); i >= 0; --i)
      {
        tmp = list;
        
        switch(arg_type) {
        case ARG_INT:
          (*((int **)field))[i + field_given] = tmp->arg.int_arg; break;
        case ARG_DOUBLE:
          (*((double **)field))[i + field_given] = tmp->arg.double_arg; break;
        case ARG_ENUM:
          (*((int **)field))[i + field_given] = tmp->arg.int_arg; break;
        case ARG_STRING:
          (*((char ***)field))[i + field_given] = tmp->arg.string_arg; break;
        default:
          break;
        }        
        (*orig_field) [i + field_given] = list->orig;
        list = list->next;
        free (tmp);
      }
  } else { /* set the default value */
    if (default_value && ! field_given) {
      switch(arg_type) {
      case ARG_INT:
      case ARG_ENUM:
        if (! *((int **)field)) {
          *((int **)field) = (int *)malloc (sizeof (int));
          (*((int **)field))[0] = default_value->int_arg; 
        }
        break;
      case ARG_DOUBLE:
        if (! *((double **)field)) {
          *((double **)field) = (double *)malloc (sizeof (double));
          (*((double **)field))[0] = default_value->double_arg;
        }
        break;
      case ARG_STRING:
        if (! *((char ***)field)) {
          *((char ***)field) = (char **)malloc (sizeof (char *));
          (*((char ***)field))[0] = gengetopt_strdup(default_value->string_arg);
        }
        break;
      default: break;
      }
      if (!(*orig_field)) {
        *orig_field = (char **) malloc (sizeof (char *));
        (*orig_field)[0] = 0;
      }
    }
  }
}

static int check_modes(
  int given1[], const char *options1[],
                       int given2[], const char *options2[])
{
  int i = 0, j = 0, errors = 0;
  
  while (given1[i] >= 0) {
    if (given1[i]) {
      while (given2[j] >= 0) {
        if (given2[j]) {
          ++errors;
          fprintf(stderr, "%s: option %s conflicts with option %s\n",
                  package_name, options1[i], options2[j]);
        }
        ++j;
      }
    }
    ++i;
  }
  
  return errors;
}

int
cmdline_parser_internal (
  int argc, char **argv, struct gengetopt_args_info *args_info,
                        struct cmdline_parser_params *params, const char *additional_error)
{
  int c;	/* Character of the parsed option.  */

  struct generic_list * targets_list = NULL;
  struct generic_list * GATKdepths_list = NULL;
  struct generic_list * GATKdepthsList_list = NULL;
  struct generic_list * excludeTargets_list = NULL;
  struct generic_list * excludeChromosomeTargets_list = NULL;
  struct generic_list * excludeSamples_list = NULL;
  struct generic_list * mergeVCF_list = NULL;
  struct generic_list * mergeVCFlist_list = NULL;
  int error = 0;
  struct gengetopt_args_info local_args_info;
  
  int override;
  int initialize;
  int check_required;
  int check_ambiguity;
  
  package_name = argv[0];
  
  override = params->override;
  initialize = params->initialize;
  check_required = params->check_required;
  check_ambiguity = params->check_ambiguity;

  if (initialize)
    cmdline_parser_init (args_info);

  cmdline_parser_init (&local_args_info);

  optarg = 0;
  optind = 0;
  opterr = params->print_errors;
  optopt = '?';

  while (1)
    {
      int option_index = 0;

      static struct option long_options[] = {
        { "help",	0, NULL, 'h' },
        { "detailed-help",	0, NULL, 0 },
        { "full-help",	0, NULL, 0 },
        { "version",	0, NULL, 'V' },
        { "referenceFASTA",	1, NULL, 'F' },
        { "readDepths",	1, NULL, 'r' },
        { "prepareTargets",	0, NULL, 'S' },
        { "targets",	1, NULL, 0 },
        { "mergedTargets",	1, NULL, 0 },
        { "mergeGATKdepths",	0, NULL, 'A' },
        { "GATKdepths",	1, NULL, 0 },
        { "GATKdepthsList",	1, NULL, 0 },
        { "sampleIDmap",	1, NULL, 0 },
        { "fromID",	1, NULL, 0 },
        { "toID",	1, NULL, 0 },
        { "columnSuffix",	1, NULL, 0 },
        { "rdPrecision",	1, NULL, 0 },
        { "outputTargetsBySamples",	0, NULL, 0 },
        { "matrix",	0, NULL, 'M' },
        { "excludeTargets",	1, NULL, 0 },
        { "excludeChromosomeTargets",	1, NULL, 0 },
        { "excludeSamples",	1, NULL, 0 },
        { "minTargetSize",	1, NULL, 0 },
        { "maxTargetSize",	1, NULL, 0 },
        { "minMeanTargetRD",	1, NULL, 0 },
        { "maxMeanTargetRD",	1, NULL, 0 },
        { "minSdTargetRD",	1, NULL, 0 },
        { "maxSdTargetRD",	1, NULL, 0 },
        { "minMeanSampleRD",	1, NULL, 0 },
        { "maxMeanSampleRD",	1, NULL, 0 },
        { "minSdSampleRD",	1, NULL, 0 },
        { "maxSdSampleRD",	1, NULL, 0 },
        { "scaleDataBySum",	0, NULL, 0 },
        { "scaleDataBySumType",	1, NULL, 0 },
        { "scaleDataBySumFactor",	1, NULL, 0 },
        { "log10",	1, NULL, 0 },
        { "centerData",	0, NULL, 0 },
        { "centerType",	1, NULL, 0 },
        { "zScoreData",	0, NULL, 0 },
        { "outputExcludedTargets",	1, NULL, 0 },
        { "outputExcludedSamples",	1, NULL, 0 },
        { "outputMatrix",	1, NULL, 'o' },
        { "PCA",	0, NULL, 'P' },
        { "PCA_saveMemory",	2, NULL, 0 },
        { "normalize",	0, NULL, 'N' },
        { "normalizeOutput",	1, NULL, 'n' },
        { "PCnormalizeMethod",	1, NULL, 0 },
        { "numPCtoRemove",	1, NULL, 0 },
        { "PVE_mean_factor",	1, NULL, 0 },
        { "PVE_contrib",	1, NULL, 0 },
        { "PCAfiles",	1, NULL, 0 },
        { "createDB",	0, NULL, 'C' },
        { "discover",	0, NULL, 'D' },
        { "xcnv",	1, NULL, 'c' },
        { "optDiscover",	0, NULL, 0 },
        { "discoverSomeQualThresh",	1, NULL, 't' },
        { "posteriorFiles",	1, NULL, 's' },
        { "genotype",	0, NULL, 'G' },
        { "gxcnv",	1, NULL, 'g' },
        { "subsegments",	0, NULL, 0 },
        { "maxTargetsInSubsegment",	1, NULL, 0 },
        { "genotypeQualThresholdWhenNoExact",	1, NULL, 'T' },
        { "mergeVCFs",	0, NULL, 0 },
        { "mergeVCF",	1, NULL, 0 },
        { "mergeVCFlist",	1, NULL, 0 },
        { "paramFile",	1, NULL, 'p' },
        { "DB",	1, NULL, 'd' },
        { "maxNormalizedReadDepthVal",	1, NULL, 'm' },
        { "maxQualScore",	1, NULL, 'q' },
        { "scorePrecision",	1, NULL, 'e' },
        { "aux_xcnv",	1, NULL, 'a' },
        { "auxUpstreamPrintTargs",	1, NULL, 'u' },
        { "auxDownstreamPrintTargs",	1, NULL, 'w' },
        { "origReadDepths",	1, NULL, 'R' },
        { "keepSampleIDs",	1, NULL, 0 },
        { "vcf",	1, NULL, 'v' },
        { "printHMM",	0, NULL, 0 },
        { "transition",	0, NULL, 0 },
        { 0,  0, 0, 0 }
      };

      c = getopt_long (argc, argv, "hVF:r:SAMo:PNn:CDc:t:s:Gg:T:p:d:m:q:e:a:u:w:R:v:", long_options, &option_index);

      if (c == -1) break;	/* Exit from `while (1)' loop.  */

      switch (c)
        {
        case 'h':	/* Print help and exit.  */
          cmdline_parser_print_help ();
          cmdline_parser_free (&local_args_info);
          exit (EXIT_SUCCESS);

        case 'V':	/* Print version and exit.  */
          cmdline_parser_print_version ();
          cmdline_parser_free (&local_args_info);
          exit (EXIT_SUCCESS);

        case 'F':	/* Reference FASTA file (MUST have .fai index file).  */
        
        
          if (update_arg( (void *)&(args_info->referenceFASTA_arg), 
               &(args_info->referenceFASTA_orig), &(args_info->referenceFASTA_given),
              &(local_args_info.referenceFASTA_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "referenceFASTA", 'F',
              additional_error))
            goto failure;
        
          break;
        case 'r':	/* Matrix of *input* read-depths, where rows (samples) and columns (targets) are labeled.  */
        
        
          if (update_arg( (void *)&(args_info->readDepths_arg), 
               &(args_info->readDepths_orig), &(args_info->readDepths_given),
              &(local_args_info.readDepths_given), optarg, 0, "-", ARG_STRING,
              check_ambiguity, override, 0, 0,
              "readDepths", 'r',
              additional_error))
            goto failure;
        
          break;
        case 'S':	/* .  */
          args_info->prepareTargets_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->prepareTargets_given),
              &(local_args_info.prepareTargets_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "prepareTargets", 'S',
              additional_error))
            goto failure;
        
          break;
        case 'A':	/* .  */
          args_info->mergeGATKdepths_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->mergeGATKdepths_given),
              &(local_args_info.mergeGATKdepths_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "mergeGATKdepths", 'A',
              additional_error))
            goto failure;
        
          break;
        case 'M':	/* .  */
          args_info->matrix_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->matrix_given),
              &(local_args_info.matrix_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "matrix", 'M',
              additional_error))
            goto failure;
        
          break;
        case 'o':	/* Read-depth matrix output file.  */
        
        
          if (update_arg( (void *)&(args_info->outputMatrix_arg), 
               &(args_info->outputMatrix_orig), &(args_info->outputMatrix_given),
              &(local_args_info.outputMatrix_given), optarg, 0, "-", ARG_STRING,
              check_ambiguity, override, 0, 0,
              "outputMatrix", 'o',
              additional_error))
            goto failure;
        
          break;
        case 'P':	/* Matrix is read from --readDepths argument; normalization factors sent to --PCAfiles argument.  */
          args_info->PCA_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->PCA_given),
              &(local_args_info.PCA_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "PCA", 'P',
              additional_error))
            goto failure;
        
          break;
        case 'N':	/* Matrix is read from --readDepths argument; normalization factors read from --PCAfiles argument.  */
          args_info->normalize_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->normalize_given),
              &(local_args_info.normalize_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "normalize", 'N',
              additional_error))
            goto failure;
        
          break;
        case 'n':	/* Normalized read-depth matrix output file.  */
          args_info->normalize_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->normalizeOutput_arg), 
               &(args_info->normalizeOutput_orig), &(args_info->normalizeOutput_given),
              &(local_args_info.normalizeOutput_given), optarg, 0, "-", ARG_STRING,
              check_ambiguity, override, 0, 0,
              "normalizeOutput", 'n',
              additional_error))
            goto failure;
        
          break;
        case 'C':	/* .  */
        
        
          if (update_arg( 0 , 
               0 , &(args_info->createDB_given),
              &(local_args_info.createDB_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "createDB", 'C',
              additional_error))
            goto failure;
        
          break;
        case 'D':	/* Matrix is read from --readDepths argument.  */
          args_info->discover_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->discover_given),
              &(local_args_info.discover_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "discover", 'D',
              additional_error))
            goto failure;
        
          break;
        case 'c':	/* CNV output file.  */
          args_info->discover_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->xcnv_arg), 
               &(args_info->xcnv_orig), &(args_info->xcnv_given),
              &(local_args_info.xcnv_given), optarg, 0, "-", ARG_STRING,
              check_ambiguity, override, 0, 0,
              "xcnv", 'c',
              additional_error))
            goto failure;
        
          break;
        case 't':	/* Quality threshold for discovering a CNV in a sample.  */
          args_info->discover_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->discoverSomeQualThresh_arg), 
               &(args_info->discoverSomeQualThresh_orig), &(args_info->discoverSomeQualThresh_given),
              &(local_args_info.discoverSomeQualThresh_given), optarg, 0, "30", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "discoverSomeQualThresh", 't',
              additional_error))
            goto failure;
        
          break;
        case 's':	/* Base file name for posterior probabilities output files; if not given, and --xcnv is not '-', this will default to --xcnv argument.  */
          args_info->discover_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->posteriorFiles_arg), 
               &(args_info->posteriorFiles_orig), &(args_info->posteriorFiles_given),
              &(local_args_info.posteriorFiles_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "posteriorFiles", 's',
              additional_error))
            goto failure;
        
          break;
        case 'G':	/* Matrix is read from --readDepths argument.  */
          args_info->genotype_mode_counter += 1;
        
        
          if (update_arg( 0 , 
               0 , &(args_info->genotype_given),
              &(local_args_info.genotype_given), optarg, 0, 0, ARG_NO,
              check_ambiguity, override, 0, 0,
              "genotype", 'G',
              additional_error))
            goto failure;
        
          break;
        case 'g':	/* xhmm CNV input file to genotype in 'readDepths' sample.  */
          args_info->genotype_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->gxcnv_arg), 
               &(args_info->gxcnv_orig), &(args_info->gxcnv_given),
              &(local_args_info.gxcnv_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "gxcnv", 'g',
              additional_error))
            goto failure;
        
          break;
        case 'T':	/* Quality threshold for calling a genotype, used *ONLY* when 'gxcnv' does not contain the 'Q_EXACT' field for the interval being genotyped.  */
          args_info->genotype_mode_counter += 1;
        
        
          if (update_arg( (void *)&(args_info->genotypeQualThresholdWhenNoExact_arg), 
               &(args_info->genotypeQualThresholdWhenNoExact_orig), &(args_info->genotypeQualThresholdWhenNoExact_given),
              &(local_args_info.genotypeQualThresholdWhenNoExact_given), optarg, 0, "20", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "genotypeQualThresholdWhenNoExact", 'T',
              additional_error))
            goto failure;
        
          break;
        case 'p':	/* (Initial) model parameters file.  */
        
        
          if (update_arg( (void *)&(args_info->paramFile_arg), 
               &(args_info->paramFile_orig), &(args_info->paramFile_given),
              &(local_args_info.paramFile_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "paramFile", 'p',
              additional_error))
            goto failure;
        
          break;
        case 'd':	/* Database storage file.  */
        
        
          if (update_arg( (void *)&(args_info->DB_arg), 
               &(args_info->DB_orig), &(args_info->DB_given),
              &(local_args_info.DB_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "DB", 'd',
              additional_error))
            goto failure;
        
          break;
        case 'm':	/* Value at which to cap the absolute value of *normalized* input read depth values ('readDepths').  */
        
        
          if (update_arg( (void *)&(args_info->maxNormalizedReadDepthVal_arg), 
               &(args_info->maxNormalizedReadDepthVal_orig), &(args_info->maxNormalizedReadDepthVal_given),
              &(local_args_info.maxNormalizedReadDepthVal_given), optarg, 0, "10", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "maxNormalizedReadDepthVal", 'm',
              additional_error))
            goto failure;
        
          break;
        case 'q':	/* Value at which to cap the calculated quality scores.  */
        
        
          if (update_arg( (void *)&(args_info->maxQualScore_arg), 
               &(args_info->maxQualScore_orig), &(args_info->maxQualScore_given),
              &(local_args_info.maxQualScore_given), optarg, 0, "99", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "maxQualScore", 'q',
              additional_error))
            goto failure;
        
          break;
        case 'e':	/* Decimal precision of quality scores.  */
        
        
          if (update_arg( (void *)&(args_info->scorePrecision_arg), 
               &(args_info->scorePrecision_orig), &(args_info->scorePrecision_given),
              &(local_args_info.scorePrecision_given), optarg, 0, "0", ARG_INT,
              check_ambiguity, override, 0, 0,
              "scorePrecision", 'e',
              additional_error))
            goto failure;
        
          break;
        case 'a':	/* Auxiliary CNV output file (may be VERY LARGE in 'genotype' mode).  */
        
        
          if (update_arg( (void *)&(args_info->aux_xcnv_arg), 
               &(args_info->aux_xcnv_orig), &(args_info->aux_xcnv_given),
              &(local_args_info.aux_xcnv_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "aux_xcnv", 'a',
              additional_error))
            goto failure;
        
          break;
        case 'u':	/* Number of targets to print upstream of CNV in 'auxOutput' file.  */
        
        
          if (update_arg( (void *)&(args_info->auxUpstreamPrintTargs_arg), 
               &(args_info->auxUpstreamPrintTargs_orig), &(args_info->auxUpstreamPrintTargs_given),
              &(local_args_info.auxUpstreamPrintTargs_given), optarg, 0, "2", ARG_INT,
              check_ambiguity, override, 0, 0,
              "auxUpstreamPrintTargs", 'u',
              additional_error))
            goto failure;
        
          break;
        case 'w':	/* Number of targets to print downstream of CNV in 'auxOutput' file.  */
        
        
          if (update_arg( (void *)&(args_info->auxDownstreamPrintTargs_arg), 
               &(args_info->auxDownstreamPrintTargs_orig), &(args_info->auxDownstreamPrintTargs_given),
              &(local_args_info.auxDownstreamPrintTargs_given), optarg, 0, "2", ARG_INT,
              check_ambiguity, override, 0, 0,
              "auxDownstreamPrintTargs", 'w',
              additional_error))
            goto failure;
        
          break;
        case 'R':	/* Matrix of unnormalized read-depths to use for CNV annotation, where rows (samples) and columns (targets) are labeled.  */
        
        
          if (update_arg( (void *)&(args_info->origReadDepths_arg), 
               &(args_info->origReadDepths_orig), &(args_info->origReadDepths_given),
              &(local_args_info.origReadDepths_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "origReadDepths", 'R',
              additional_error))
            goto failure;
        
          break;
        case 'v':	/* Genotyped CNV output VCF file.  */
        
        
          if (update_arg( (void *)&(args_info->vcf_arg), 
               &(args_info->vcf_orig), &(args_info->vcf_given),
              &(local_args_info.vcf_given), optarg, 0, "-", ARG_STRING,
              check_ambiguity, override, 0, 0,
              "vcf", 'v',
              additional_error))
            goto failure;
        
          break;

        case 0:	/* Long option with no short option */
          if (strcmp (long_options[option_index].name, "detailed-help") == 0) {
            cmdline_parser_print_detailed_help ();
            cmdline_parser_free (&local_args_info);
            exit (EXIT_SUCCESS);
          }

          if (strcmp (long_options[option_index].name, "full-help") == 0) {
            cmdline_parser_print_full_help ();
            cmdline_parser_free (&local_args_info);
            exit (EXIT_SUCCESS);
          }

          /* Input targets lists.  */
          if (strcmp (long_options[option_index].name, "targets") == 0)
          {
            args_info->prepareTargets_mode_counter += 1;
          
            if (update_multiple_arg_temp(&targets_list, 
                &(local_args_info.targets_given), optarg, 0, 0, ARG_STRING,
                "targets", '-',
                additional_error))
              goto failure;
          
          }
          /* Output targets list.  */
          else if (strcmp (long_options[option_index].name, "mergedTargets") == 0)
          {
            args_info->prepareTargets_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->mergedTargets_arg), 
                 &(args_info->mergedTargets_orig), &(args_info->mergedTargets_given),
                &(local_args_info.mergedTargets_given), optarg, 0, "-", ARG_STRING,
                check_ambiguity, override, 0, 0,
                "mergedTargets", '-',
                additional_error))
              goto failure;
          
          }
          /* GATK sample_interval_summary output file(s) to be merged [must have *IDENTICAL* target lists].  */
          else if (strcmp (long_options[option_index].name, "GATKdepths") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
            if (update_multiple_arg_temp(&GATKdepths_list, 
                &(local_args_info.GATKdepths_given), optarg, 0, 0, ARG_STRING,
                "GATKdepths", '-',
                additional_error))
              goto failure;
          
          }
          /* A file containing a list of GATK sample_interval_summary output files to be merged [must have *IDENTICAL* target lists].  */
          else if (strcmp (long_options[option_index].name, "GATKdepthsList") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
            if (update_multiple_arg_temp(&GATKdepthsList_list, 
                &(local_args_info.GATKdepthsList_given), optarg, 0, 0, ARG_STRING,
                "GATKdepthsList", '-',
                additional_error))
              goto failure;
          
          }
          /* File containing mappings of sample names to new sample names (in columns designated by fromID, toID).  */
          else if (strcmp (long_options[option_index].name, "sampleIDmap") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->sampleIDmap_arg), 
                 &(args_info->sampleIDmap_orig), &(args_info->sampleIDmap_given),
                &(local_args_info.sampleIDmap_given), optarg, 0, 0, ARG_STRING,
                check_ambiguity, override, 0, 0,
                "sampleIDmap", '-',
                additional_error))
              goto failure;
          
          }
          /* Column number of OLD sample IDs to map.  */
          else if (strcmp (long_options[option_index].name, "fromID") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->fromID_arg), 
                 &(args_info->fromID_orig), &(args_info->fromID_given),
                &(local_args_info.fromID_given), optarg, 0, "1", ARG_INT,
                check_ambiguity, override, 0, 0,
                "fromID", '-',
                additional_error))
              goto failure;
          
          }
          /* Column number of NEW sample IDs to map.  */
          else if (strcmp (long_options[option_index].name, "toID") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->toID_arg), 
                 &(args_info->toID_orig), &(args_info->toID_given),
                &(local_args_info.toID_given), optarg, 0, "2", ARG_INT,
                check_ambiguity, override, 0, 0,
                "toID", '-',
                additional_error))
              goto failure;
          
          }
          /* Suffix of columns to be used for merging [where columns are in the form: SAMPLE + columnSuffix].  */
          else if (strcmp (long_options[option_index].name, "columnSuffix") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->columnSuffix_arg), 
                 &(args_info->columnSuffix_orig), &(args_info->columnSuffix_given),
                &(local_args_info.columnSuffix_given), optarg, 0, "_mean_cvg", ARG_STRING,
                check_ambiguity, override, 0, 0,
                "columnSuffix", '-',
                additional_error))
              goto failure;
          
          }
          /* Decimal precision of read depths output.  */
          else if (strcmp (long_options[option_index].name, "rdPrecision") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->rdPrecision_arg), 
                 &(args_info->rdPrecision_orig), &(args_info->rdPrecision_given),
                &(local_args_info.rdPrecision_given), optarg, 0, "2", ARG_INT,
                check_ambiguity, override, 0, 0,
                "rdPrecision", '-',
                additional_error))
              goto failure;
          
          }
          /* Output targets x samples (instead of samples x targets).  */
          else if (strcmp (long_options[option_index].name, "outputTargetsBySamples") == 0)
          {
            args_info->mergeGATKdepths_mode_counter += 1;
          
          
            if (update_arg((void *)&(args_info->outputTargetsBySamples_flag), 0, &(args_info->outputTargetsBySamples_given),
                &(local_args_info.outputTargetsBySamples_given), optarg, 0, 0, ARG_FLAG,
                check_ambiguity, override, 1, 0, "outputTargetsBySamples", '-',
                additional_error))
              goto failure;
          
          }
          /* File(s) of targets to exclude.  */
          else if (strcmp (long_options[option_index].name, "excludeTargets") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
            if (update_multiple_arg_temp(&excludeTargets_list, 
                &(local_args_info.excludeTargets_given), optarg, 0, 0, ARG_STRING,
                "excludeTargets", '-',
                additional_error))
              goto failure;
          
          }
          /* Target chromosome(s) to exclude.  */
          else if (strcmp (long_options[option_index].name, "excludeChromosomeTargets") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
            if (update_multiple_arg_temp(&excludeChromosomeTargets_list, 
                &(local_args_info.excludeChromosomeTargets_given), optarg, 0, 0, ARG_STRING,
                "excludeChromosomeTargets", '-',
                additional_error))
              goto failure;
          
          }
          /* File(s) of samples to exclude.  */
          else if (strcmp (long_options[option_index].name, "excludeSamples") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
            if (update_multiple_arg_temp(&excludeSamples_list, 
                &(local_args_info.excludeSamples_given), optarg, 0, 0, ARG_STRING,
                "excludeSamples", '-',
                additional_error))
              goto failure;
          
          }
          /* Minimum size of target (in bp) to process.  */
          else if (strcmp (long_options[option_index].name, "minTargetSize") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->minTargetSize_arg), 
                 &(args_info->minTargetSize_orig), &(args_info->minTargetSize_given),
                &(local_args_info.minTargetSize_given), optarg, 0, "0", ARG_INT,
                check_ambiguity, override, 0, 0,
                "minTargetSize", '-',
                additional_error))
              goto failure;
          
          }
          /* Maximum size of target (in bp) to process.  */
          else if (strcmp (long_options[option_index].name, "maxTargetSize") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxTargetSize_arg), 
                 &(args_info->maxTargetSize_orig), &(args_info->maxTargetSize_given),
                &(local_args_info.maxTargetSize_given), optarg, 0, 0, ARG_INT,
                check_ambiguity, override, 0, 0,
                "maxTargetSize", '-',
                additional_error))
              goto failure;
          
          }
          /* Minimum per-target mean RD to require for target to be processed.  */
          else if (strcmp (long_options[option_index].name, "minMeanTargetRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->minMeanTargetRD_arg), 
                 &(args_info->minMeanTargetRD_orig), &(args_info->minMeanTargetRD_given),
                &(local_args_info.minMeanTargetRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "minMeanTargetRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Maximum per-target mean RD to require for target to be processed.  */
          else if (strcmp (long_options[option_index].name, "maxMeanTargetRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxMeanTargetRD_arg), 
                 &(args_info->maxMeanTargetRD_orig), &(args_info->maxMeanTargetRD_given),
                &(local_args_info.maxMeanTargetRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "maxMeanTargetRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Minimum per-target standard deviation of RD to require for target to be processed.  */
          else if (strcmp (long_options[option_index].name, "minSdTargetRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->minSdTargetRD_arg), 
                 &(args_info->minSdTargetRD_orig), &(args_info->minSdTargetRD_given),
                &(local_args_info.minSdTargetRD_given), optarg, 0, "0", ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "minSdTargetRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Maximum per-target standard deviation of RD to require for target to be processed.  */
          else if (strcmp (long_options[option_index].name, "maxSdTargetRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxSdTargetRD_arg), 
                 &(args_info->maxSdTargetRD_orig), &(args_info->maxSdTargetRD_given),
                &(local_args_info.maxSdTargetRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "maxSdTargetRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Minimum per-sample mean RD to require for sample to be processed.  */
          else if (strcmp (long_options[option_index].name, "minMeanSampleRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->minMeanSampleRD_arg), 
                 &(args_info->minMeanSampleRD_orig), &(args_info->minMeanSampleRD_given),
                &(local_args_info.minMeanSampleRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "minMeanSampleRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Maximum per-sample mean RD to require for sample to be processed.  */
          else if (strcmp (long_options[option_index].name, "maxMeanSampleRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxMeanSampleRD_arg), 
                 &(args_info->maxMeanSampleRD_orig), &(args_info->maxMeanSampleRD_given),
                &(local_args_info.maxMeanSampleRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "maxMeanSampleRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Minimum per-sample standard deviation of RD to require for sample to be processed.  */
          else if (strcmp (long_options[option_index].name, "minSdSampleRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->minSdSampleRD_arg), 
                 &(args_info->minSdSampleRD_orig), &(args_info->minSdSampleRD_given),
                &(local_args_info.minSdSampleRD_given), optarg, 0, "0", ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "minSdSampleRD", '-',
                additional_error))
              goto failure;
          
          }
          /* Maximum per-sample standard deviation of RD to require for sample to be processed.  */
          else if (strcmp (long_options[option_index].name, "maxSdSampleRD") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxSdSampleRD_arg), 
                 &(args_info->maxSdSampleRD_orig), &(args_info->maxSdSampleRD_given),
                &(local_args_info.maxSdSampleRD_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "maxSdSampleRD", '-',
                additional_error))
              goto failure;
          
          }
          /* After any filtering, scale read-depth matrix values by sample- or target- sums (as per --scaleDataBySumType) [i.e., divide by row or column sums], but multiply by factor specificied by --scaleDataBySumFactor.  */
          else if (strcmp (long_options[option_index].name, "scaleDataBySum") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg((void *)&(args_info->scaleDataBySum_flag), 0, &(args_info->scaleDataBySum_given),
                &(local_args_info.scaleDataBySum_given), optarg, 0, 0, ARG_FLAG,
                check_ambiguity, override, 1, 0, "scaleDataBySum", '-',
                additional_error))
              goto failure;
          
          }
          /* If --scaleDataBySum given, then scale the data within this dimension.  */
          else if (strcmp (long_options[option_index].name, "scaleDataBySumType") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->scaleDataBySumType_arg), 
                 &(args_info->scaleDataBySumType_orig), &(args_info->scaleDataBySumType_given),
                &(local_args_info.scaleDataBySumType_given), optarg, cmdline_parser_scaleDataBySumType_values, 0, ARG_ENUM,
                check_ambiguity, override, 0, 0,
                "scaleDataBySumType", '-',
                additional_error))
              goto failure;
          
          }
          /* If --scaleDataBySum given, then divide by appropriate sum (but multiply by this factor).  */
          else if (strcmp (long_options[option_index].name, "scaleDataBySumFactor") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->scaleDataBySumFactor_arg), 
                 &(args_info->scaleDataBySumFactor_orig), &(args_info->scaleDataBySumFactor_given),
                &(local_args_info.scaleDataBySumFactor_given), optarg, 0, "1e6", ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "scaleDataBySumFactor", '-',
                additional_error))
              goto failure;
          
          }
          /* After any filtering and optional scaling steps (but before any optional centering steps), convert the matrix to log10 values. To deal with non-positive and small positive values, a truncation and then pseudocount approach is taken. Specifically, denote the original matrix value as x and this parameter's pseudocount value as v (say, 0.5). The matrix value used is then log10(max(x, 0) + v).  */
          else if (strcmp (long_options[option_index].name, "log10") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->log10_arg), 
                 &(args_info->log10_orig), &(args_info->log10_given),
                &(local_args_info.log10_given), optarg, 0, 0, ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "log10", '-',
                additional_error))
              goto failure;
          
          }
          /* Output sample- or target- centered read-depth matrix (as per --centerType).  */
          else if (strcmp (long_options[option_index].name, "centerData") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg((void *)&(args_info->centerData_flag), 0, &(args_info->centerData_given),
                &(local_args_info.centerData_given), optarg, 0, 0, ARG_FLAG,
                check_ambiguity, override, 1, 0, "centerData", '-',
                additional_error))
              goto failure;
          
          }
          /* If --centerData given, then center the data around this dimension.  */
          else if (strcmp (long_options[option_index].name, "centerType") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->centerType_arg), 
                 &(args_info->centerType_orig), &(args_info->centerType_given),
                &(local_args_info.centerType_given), optarg, cmdline_parser_centerType_values, 0, ARG_ENUM,
                check_ambiguity, override, 0, 0,
                "centerType", '-',
                additional_error))
              goto failure;
          
          }
          /* If --centerData given, then additionally normalize by standard deviation (outputting z-scores).  */
          else if (strcmp (long_options[option_index].name, "zScoreData") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg((void *)&(args_info->zScoreData_flag), 0, &(args_info->zScoreData_given),
                &(local_args_info.zScoreData_given), optarg, 0, 0, ARG_FLAG,
                check_ambiguity, override, 1, 0, "zScoreData", '-',
                additional_error))
              goto failure;
          
          }
          /* File in which to output targets excluded by some criterion.  */
          else if (strcmp (long_options[option_index].name, "outputExcludedTargets") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->outputExcludedTargets_arg), 
                 &(args_info->outputExcludedTargets_orig), &(args_info->outputExcludedTargets_given),
                &(local_args_info.outputExcludedTargets_given), optarg, 0, 0, ARG_STRING,
                check_ambiguity, override, 0, 0,
                "outputExcludedTargets", '-',
                additional_error))
              goto failure;
          
          }
          /* File in which to output samples excluded by some criterion.  */
          else if (strcmp (long_options[option_index].name, "outputExcludedSamples") == 0)
          {
            args_info->matrix_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->outputExcludedSamples_arg), 
                 &(args_info->outputExcludedSamples_orig), &(args_info->outputExcludedSamples_given),
                &(local_args_info.outputExcludedSamples_given), optarg, 0, 0, ARG_STRING,
                check_ambiguity, override, 0, 0,
                "outputExcludedSamples", '-',
                additional_error))
              goto failure;
          
          }
          /* Should XHMM save memory by storing some of the intermediate PCA matrices as temporary files on disk?.  */
          else if (strcmp (long_options[option_index].name, "PCA_saveMemory") == 0)
          {
            args_info->PCA_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->PCA_saveMemory_arg), 
                 &(args_info->PCA_saveMemory_orig), &(args_info->PCA_saveMemory_given),
                &(local_args_info.PCA_saveMemory_given), optarg, 0, "", ARG_STRING,
                check_ambiguity, override, 0, 0,
                "PCA_saveMemory", '-',
                additional_error))
              goto failure;
          
          }
          /* Method to choose which prinicipal components are removed for data normalization.  */
          else if (strcmp (long_options[option_index].name, "PCnormalizeMethod") == 0)
          {
            args_info->normalize_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->PCnormalizeMethod_arg), 
                 &(args_info->PCnormalizeMethod_orig), &(args_info->PCnormalizeMethod_given),
                &(local_args_info.PCnormalizeMethod_given), optarg, cmdline_parser_PCnormalizeMethod_values, "PVE_mean", ARG_ENUM,
                check_ambiguity, override, 0, 0,
                "PCnormalizeMethod", '-',
                additional_error))
              goto failure;
          
          }
          /* Number of highest principal components to filter out.  */
          else if (strcmp (long_options[option_index].name, "numPCtoRemove") == 0)
          {
            args_info->normalize_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->numPCtoRemove_arg), 
                 &(args_info->numPCtoRemove_orig), &(args_info->numPCtoRemove_given),
                &(local_args_info.numPCtoRemove_given), optarg, 0, "20", ARG_INT,
                check_ambiguity, override, 0, 0,
                "numPCtoRemove", '-',
                additional_error))
              goto failure;
          
          }
          /* Remove all principal components that individually explain more variance than this factor times the average (in the original PCA-ed data).  */
          else if (strcmp (long_options[option_index].name, "PVE_mean_factor") == 0)
          {
            args_info->normalize_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->PVE_mean_factor_arg), 
                 &(args_info->PVE_mean_factor_orig), &(args_info->PVE_mean_factor_given),
                &(local_args_info.PVE_mean_factor_given), optarg, 0, "0.7", ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "PVE_mean_factor", '-',
                additional_error))
              goto failure;
          
          }
          /* Remove the smallest number of principal components that explain this percent of the variance (in the original PCA-ed data).  */
          else if (strcmp (long_options[option_index].name, "PVE_contrib") == 0)
          {
            args_info->normalize_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->PVE_contrib_arg), 
                 &(args_info->PVE_contrib_orig), &(args_info->PVE_contrib_given),
                &(local_args_info.PVE_contrib_given), optarg, 0, "50", ARG_DOUBLE,
                check_ambiguity, override, 0, 0,
                "PVE_contrib", '-',
                additional_error))
              goto failure;
          
          }
          /* Base file name for 'PCA' *output*, and 'normalize' *input*.  */
          else if (strcmp (long_options[option_index].name, "PCAfiles") == 0)
          {
          
          
            if (update_arg( (void *)&(args_info->PCAfiles_arg), 
                 &(args_info->PCAfiles_orig), &(args_info->PCAfiles_given),
                &(local_args_info.PCAfiles_given), optarg, 0, 0, ARG_STRING,
                check_ambiguity, override, 0, 0,
                "PCAfiles", '-',
                additional_error))
              goto failure;
          
          }
          /* Optimize input HMM parameters before discovering CNVs.  */
          else if (strcmp (long_options[option_index].name, "optDiscover") == 0)
          {
            args_info->discover_mode_counter += 1;
          
          
            if (update_arg( 0 , 
                 0 , &(args_info->optDiscover_given),
                &(local_args_info.optDiscover_given), optarg, 0, 0, ARG_NO,
                check_ambiguity, override, 0, 0,
                "optDiscover", '-',
                additional_error))
              goto failure;
          
          }
          /* In addition to genotyping the intervals specified in gxcnv, genotype all sub-segments of these intervals (with maxTargetsInSubsegment or fewer targets).  */
          else if (strcmp (long_options[option_index].name, "subsegments") == 0)
          {
            args_info->genotype_mode_counter += 1;
          
          
            if (update_arg((void *)&(args_info->subsegments_flag), 0, &(args_info->subsegments_given),
                &(local_args_info.subsegments_given), optarg, 0, 0, ARG_FLAG,
                check_ambiguity, override, 1, 0, "subsegments", '-',
                additional_error))
              goto failure;
          
          }
          /* When genotyping sub-segments of input intervals, only consider sub-segments consisting of this number of targets or fewer.  */
          else if (strcmp (long_options[option_index].name, "maxTargetsInSubsegment") == 0)
          {
            args_info->genotype_mode_counter += 1;
          
          
            if (update_arg( (void *)&(args_info->maxTargetsInSubsegment_arg), 
                 &(args_info->maxTargetsInSubsegment_orig), &(args_info->maxTargetsInSubsegment_given),
                &(local_args_info.maxTargetsInSubsegment_given), optarg, 0, "30", ARG_INT,
                check_ambiguity, override, 0, 0,
                "maxTargetsInSubsegment", '-',
                additional_error))
              goto failure;
          
          }
          /* .  */
          else if (strcmp (long_options[option_index].name, "mergeVCFs") == 0)
          {
            args_info->mergeVCFs_mode_counter += 1;
          
          
            if (update_arg( 0 , 
                 0 , &(args_info->mergeVCFs_given),
                &(local_args_info.mergeVCFs_given), optarg, 0, 0, ARG_NO,
                check_ambiguity, override, 0, 0,
                "mergeVCFs", '-',
                additional_error))
              goto failure;
          
          }
          /* VCF file(s) to be merged [must have *IDENTICAL* genotyped intervals].  */
          else if (strcmp (long_options[option_index].name, "mergeVCF") == 0)
          {
            args_info->mergeVCFs_mode_counter += 1;
          
            if (update_multiple_arg_temp(&mergeVCF_list, 
                &(local_args_info.mergeVCF_given), optarg, 0, 0, ARG_STRING,
                "mergeVCF", '-',
                additional_error))
              goto failure;
          
          }
          /* A file containing a list of VCF files to be merged [must have *IDENTICAL* genotyped intervals].  */
          else if (strcmp (long_options[option_index].name, "mergeVCFlist") == 0)
          {
            args_info->mergeVCFs_mode_counter += 1;
          
            if (update_multiple_arg_temp(&mergeVCFlist_list, 
                &(local_args_info.mergeVCFlist_given), optarg, 0, 0, ARG_STRING,
                "mergeVCFlist", '-',
                additional_error))
              goto failure;
          
          }
          /* File containing a list of samples to be analyzed.  */
          else if (strcmp (long_options[option_index].name, "keepSampleIDs") == 0)
          {
          
          
            if (update_arg( (void *)&(args_info->keepSampleIDs_arg), 
                 &(args_info->keepSampleIDs_orig), &(args_info->keepSampleIDs_given),
                &(local_args_info.keepSampleIDs_given), optarg, 0, 0, ARG_STRING,
                check_ambiguity, override, 0, 0,
                "keepSampleIDs", '-',
                additional_error))
              goto failure;
          
          }
          /* .  */
          else if (strcmp (long_options[option_index].name, "printHMM") == 0)
          {
            args_info->printHMM_mode_counter += 1;
          
          
            if (update_arg( 0 , 
                 0 , &(args_info->printHMM_given),
                &(local_args_info.printHMM_given), optarg, 0, 0, ARG_NO,
                check_ambiguity, override, 0, 0,
                "printHMM", '-',
                additional_error))
              goto failure;
          
          }
          /* .  */
          else if (strcmp (long_options[option_index].name, "transition") == 0)
          {
            args_info->transition_mode_counter += 1;
          
          
            if (update_arg( 0 , 
                 0 , &(args_info->transition_given),
                &(local_args_info.transition_given), optarg, 0, 0, ARG_NO,
                check_ambiguity, override, 0, 0,
                "transition", '-',
                additional_error))
              goto failure;
          
          }
          
          break;
        case '?':	/* Invalid option.  */
          /* `getopt_long' already printed an error message.  */
          goto failure;

        default:	/* bug: option not considered.  */
          fprintf (stderr, "%s: option unknown: %c%s\n", CMDLINE_PARSER_PACKAGE, c, (additional_error ? additional_error : ""));
          abort ();
        } /* switch */
    } /* while */


  update_multiple_arg((void *)&(args_info->targets_arg),
    &(args_info->targets_orig), args_info->targets_given,
    local_args_info.targets_given, 0,
    ARG_STRING, targets_list);
  update_multiple_arg((void *)&(args_info->GATKdepths_arg),
    &(args_info->GATKdepths_orig), args_info->GATKdepths_given,
    local_args_info.GATKdepths_given, 0,
    ARG_STRING, GATKdepths_list);
  update_multiple_arg((void *)&(args_info->GATKdepthsList_arg),
    &(args_info->GATKdepthsList_orig), args_info->GATKdepthsList_given,
    local_args_info.GATKdepthsList_given, 0,
    ARG_STRING, GATKdepthsList_list);
  update_multiple_arg((void *)&(args_info->excludeTargets_arg),
    &(args_info->excludeTargets_orig), args_info->excludeTargets_given,
    local_args_info.excludeTargets_given, 0,
    ARG_STRING, excludeTargets_list);
  update_multiple_arg((void *)&(args_info->excludeChromosomeTargets_arg),
    &(args_info->excludeChromosomeTargets_orig), args_info->excludeChromosomeTargets_given,
    local_args_info.excludeChromosomeTargets_given, 0,
    ARG_STRING, excludeChromosomeTargets_list);
  update_multiple_arg((void *)&(args_info->excludeSamples_arg),
    &(args_info->excludeSamples_orig), args_info->excludeSamples_given,
    local_args_info.excludeSamples_given, 0,
    ARG_STRING, excludeSamples_list);
  update_multiple_arg((void *)&(args_info->mergeVCF_arg),
    &(args_info->mergeVCF_orig), args_info->mergeVCF_given,
    local_args_info.mergeVCF_given, 0,
    ARG_STRING, mergeVCF_list);
  update_multiple_arg((void *)&(args_info->mergeVCFlist_arg),
    &(args_info->mergeVCFlist_orig), args_info->mergeVCFlist_given,
    local_args_info.mergeVCFlist_given, 0,
    ARG_STRING, mergeVCFlist_list);

  args_info->targets_given += local_args_info.targets_given;
  local_args_info.targets_given = 0;
  args_info->GATKdepths_given += local_args_info.GATKdepths_given;
  local_args_info.GATKdepths_given = 0;
  args_info->GATKdepthsList_given += local_args_info.GATKdepthsList_given;
  local_args_info.GATKdepthsList_given = 0;
  args_info->excludeTargets_given += local_args_info.excludeTargets_given;
  local_args_info.excludeTargets_given = 0;
  args_info->excludeChromosomeTargets_given += local_args_info.excludeChromosomeTargets_given;
  local_args_info.excludeChromosomeTargets_given = 0;
  args_info->excludeSamples_given += local_args_info.excludeSamples_given;
  local_args_info.excludeSamples_given = 0;
  args_info->mergeVCF_given += local_args_info.mergeVCF_given;
  local_args_info.mergeVCF_given = 0;
  args_info->mergeVCFlist_given += local_args_info.mergeVCFlist_given;
  local_args_info.mergeVCFlist_given = 0;
  
  if (args_info->PCA_mode_counter && args_info->discover_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    error += check_modes(PCA_given, PCA_desc, discover_given, discover_desc);
  }
  if (args_info->PCA_mode_counter && args_info->genotype_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    error += check_modes(PCA_given, PCA_desc, genotype_given, genotype_desc);
  }
  if (args_info->PCA_mode_counter && args_info->matrix_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    error += check_modes(PCA_given, PCA_desc, matrix_given, matrix_desc);
  }
  if (args_info->PCA_mode_counter && args_info->mergeGATKdepths_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    error += check_modes(PCA_given, PCA_desc, mergeGATKdepths_given, mergeGATKdepths_desc);
  }
  if (args_info->PCA_mode_counter && args_info->mergeVCFs_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    error += check_modes(PCA_given, PCA_desc, mergeVCFs_given, mergeVCFs_desc);
  }
  if (args_info->PCA_mode_counter && args_info->normalize_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(PCA_given, PCA_desc, normalize_given, normalize_desc);
  }
  if (args_info->PCA_mode_counter && args_info->prepareTargets_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(PCA_given, PCA_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->PCA_mode_counter && args_info->printHMM_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(PCA_given, PCA_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->PCA_mode_counter && args_info->transition_mode_counter) {
    int PCA_given[] = {args_info->PCA_given, args_info->PCA_saveMemory_given,  -1};
    const char *PCA_desc[] = {"--PCA", "--PCA_saveMemory",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(PCA_given, PCA_desc, transition_given, transition_desc);
  }
  if (args_info->discover_mode_counter && args_info->genotype_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    error += check_modes(discover_given, discover_desc, genotype_given, genotype_desc);
  }
  if (args_info->discover_mode_counter && args_info->matrix_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    error += check_modes(discover_given, discover_desc, matrix_given, matrix_desc);
  }
  if (args_info->discover_mode_counter && args_info->mergeGATKdepths_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    error += check_modes(discover_given, discover_desc, mergeGATKdepths_given, mergeGATKdepths_desc);
  }
  if (args_info->discover_mode_counter && args_info->mergeVCFs_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    error += check_modes(discover_given, discover_desc, mergeVCFs_given, mergeVCFs_desc);
  }
  if (args_info->discover_mode_counter && args_info->normalize_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(discover_given, discover_desc, normalize_given, normalize_desc);
  }
  if (args_info->discover_mode_counter && args_info->prepareTargets_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(discover_given, discover_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->discover_mode_counter && args_info->printHMM_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(discover_given, discover_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->discover_mode_counter && args_info->transition_mode_counter) {
    int discover_given[] = {args_info->discover_given, args_info->xcnv_given, args_info->optDiscover_given, args_info->discoverSomeQualThresh_given, args_info->posteriorFiles_given,  -1};
    const char *discover_desc[] = {"--discover", "--xcnv", "--optDiscover", "--discoverSomeQualThresh", "--posteriorFiles",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(discover_given, discover_desc, transition_given, transition_desc);
  }
  if (args_info->genotype_mode_counter && args_info->matrix_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    error += check_modes(genotype_given, genotype_desc, matrix_given, matrix_desc);
  }
  if (args_info->genotype_mode_counter && args_info->mergeGATKdepths_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    error += check_modes(genotype_given, genotype_desc, mergeGATKdepths_given, mergeGATKdepths_desc);
  }
  if (args_info->genotype_mode_counter && args_info->mergeVCFs_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    error += check_modes(genotype_given, genotype_desc, mergeVCFs_given, mergeVCFs_desc);
  }
  if (args_info->genotype_mode_counter && args_info->normalize_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(genotype_given, genotype_desc, normalize_given, normalize_desc);
  }
  if (args_info->genotype_mode_counter && args_info->prepareTargets_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(genotype_given, genotype_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->genotype_mode_counter && args_info->printHMM_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(genotype_given, genotype_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->genotype_mode_counter && args_info->transition_mode_counter) {
    int genotype_given[] = {args_info->genotype_given, args_info->gxcnv_given, args_info->subsegments_given, args_info->maxTargetsInSubsegment_given, args_info->genotypeQualThresholdWhenNoExact_given,  -1};
    const char *genotype_desc[] = {"--genotype", "--gxcnv", "--subsegments", "--maxTargetsInSubsegment", "--genotypeQualThresholdWhenNoExact",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(genotype_given, genotype_desc, transition_given, transition_desc);
  }
  if (args_info->matrix_mode_counter && args_info->mergeGATKdepths_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    error += check_modes(matrix_given, matrix_desc, mergeGATKdepths_given, mergeGATKdepths_desc);
  }
  if (args_info->matrix_mode_counter && args_info->mergeVCFs_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    error += check_modes(matrix_given, matrix_desc, mergeVCFs_given, mergeVCFs_desc);
  }
  if (args_info->matrix_mode_counter && args_info->normalize_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(matrix_given, matrix_desc, normalize_given, normalize_desc);
  }
  if (args_info->matrix_mode_counter && args_info->prepareTargets_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(matrix_given, matrix_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->matrix_mode_counter && args_info->printHMM_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(matrix_given, matrix_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->matrix_mode_counter && args_info->transition_mode_counter) {
    int matrix_given[] = {args_info->matrix_given, args_info->excludeTargets_given, args_info->excludeChromosomeTargets_given, args_info->excludeSamples_given, args_info->minTargetSize_given, args_info->maxTargetSize_given, args_info->minMeanTargetRD_given, args_info->maxMeanTargetRD_given, args_info->minSdTargetRD_given, args_info->maxSdTargetRD_given, args_info->minMeanSampleRD_given, args_info->maxMeanSampleRD_given, args_info->minSdSampleRD_given, args_info->maxSdSampleRD_given, args_info->scaleDataBySum_given, args_info->scaleDataBySumType_given, args_info->scaleDataBySumFactor_given, args_info->log10_given, args_info->centerData_given, args_info->centerType_given, args_info->zScoreData_given, args_info->outputExcludedTargets_given, args_info->outputExcludedSamples_given,  -1};
    const char *matrix_desc[] = {"--matrix", "--excludeTargets", "--excludeChromosomeTargets", "--excludeSamples", "--minTargetSize", "--maxTargetSize", "--minMeanTargetRD", "--maxMeanTargetRD", "--minSdTargetRD", "--maxSdTargetRD", "--minMeanSampleRD", "--maxMeanSampleRD", "--minSdSampleRD", "--maxSdSampleRD", "--scaleDataBySum", "--scaleDataBySumType", "--scaleDataBySumFactor", "--log10", "--centerData", "--centerType", "--zScoreData", "--outputExcludedTargets", "--outputExcludedSamples",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(matrix_given, matrix_desc, transition_given, transition_desc);
  }
  if (args_info->mergeGATKdepths_mode_counter && args_info->mergeVCFs_mode_counter) {
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    error += check_modes(mergeGATKdepths_given, mergeGATKdepths_desc, mergeVCFs_given, mergeVCFs_desc);
  }
  if (args_info->mergeGATKdepths_mode_counter && args_info->normalize_mode_counter) {
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(mergeGATKdepths_given, mergeGATKdepths_desc, normalize_given, normalize_desc);
  }
  if (args_info->mergeGATKdepths_mode_counter && args_info->prepareTargets_mode_counter) {
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(mergeGATKdepths_given, mergeGATKdepths_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->mergeGATKdepths_mode_counter && args_info->printHMM_mode_counter) {
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(mergeGATKdepths_given, mergeGATKdepths_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->mergeGATKdepths_mode_counter && args_info->transition_mode_counter) {
    int mergeGATKdepths_given[] = {args_info->mergeGATKdepths_given, args_info->GATKdepths_given, args_info->GATKdepthsList_given, args_info->sampleIDmap_given, args_info->fromID_given, args_info->toID_given, args_info->columnSuffix_given, args_info->rdPrecision_given, args_info->outputTargetsBySamples_given,  -1};
    const char *mergeGATKdepths_desc[] = {"--mergeGATKdepths", "--GATKdepths", "--GATKdepthsList", "--sampleIDmap", "--fromID", "--toID", "--columnSuffix", "--rdPrecision", "--outputTargetsBySamples",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(mergeGATKdepths_given, mergeGATKdepths_desc, transition_given, transition_desc);
  }
  if (args_info->mergeVCFs_mode_counter && args_info->normalize_mode_counter) {
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    error += check_modes(mergeVCFs_given, mergeVCFs_desc, normalize_given, normalize_desc);
  }
  if (args_info->mergeVCFs_mode_counter && args_info->prepareTargets_mode_counter) {
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(mergeVCFs_given, mergeVCFs_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->mergeVCFs_mode_counter && args_info->printHMM_mode_counter) {
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(mergeVCFs_given, mergeVCFs_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->mergeVCFs_mode_counter && args_info->transition_mode_counter) {
    int mergeVCFs_given[] = {args_info->mergeVCFs_given, args_info->mergeVCF_given, args_info->mergeVCFlist_given,  -1};
    const char *mergeVCFs_desc[] = {"--mergeVCFs", "--mergeVCF", "--mergeVCFlist",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(mergeVCFs_given, mergeVCFs_desc, transition_given, transition_desc);
  }
  if (args_info->normalize_mode_counter && args_info->prepareTargets_mode_counter) {
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    error += check_modes(normalize_given, normalize_desc, prepareTargets_given, prepareTargets_desc);
  }
  if (args_info->normalize_mode_counter && args_info->printHMM_mode_counter) {
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(normalize_given, normalize_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->normalize_mode_counter && args_info->transition_mode_counter) {
    int normalize_given[] = {args_info->normalize_given, args_info->normalizeOutput_given, args_info->PCnormalizeMethod_given, args_info->numPCtoRemove_given, args_info->PVE_mean_factor_given, args_info->PVE_contrib_given,  -1};
    const char *normalize_desc[] = {"--normalize", "--normalizeOutput", "--PCnormalizeMethod", "--numPCtoRemove", "--PVE_mean_factor", "--PVE_contrib",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(normalize_given, normalize_desc, transition_given, transition_desc);
  }
  if (args_info->prepareTargets_mode_counter && args_info->printHMM_mode_counter) {
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    error += check_modes(prepareTargets_given, prepareTargets_desc, printHMM_given, printHMM_desc);
  }
  if (args_info->prepareTargets_mode_counter && args_info->transition_mode_counter) {
    int prepareTargets_given[] = {args_info->prepareTargets_given, args_info->targets_given, args_info->mergedTargets_given,  -1};
    const char *prepareTargets_desc[] = {"--prepareTargets", "--targets", "--mergedTargets",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(prepareTargets_given, prepareTargets_desc, transition_given, transition_desc);
  }
  if (args_info->printHMM_mode_counter && args_info->transition_mode_counter) {
    int printHMM_given[] = {args_info->printHMM_given,  -1};
    const char *printHMM_desc[] = {"--printHMM",  0};
    int transition_given[] = {args_info->transition_given,  -1};
    const char *transition_desc[] = {"--transition",  0};
    error += check_modes(printHMM_given, printHMM_desc, transition_given, transition_desc);
  }
  
  if (check_required)
    {
      error += cmdline_parser_required2 (args_info, argv[0], additional_error);
    }

  cmdline_parser_release (&local_args_info);

  if ( error )
    return (EXIT_FAILURE);

  return 0;

failure:
  free_list (targets_list, 1 );
  free_list (GATKdepths_list, 1 );
  free_list (GATKdepthsList_list, 1 );
  free_list (excludeTargets_list, 1 );
  free_list (excludeChromosomeTargets_list, 1 );
  free_list (excludeSamples_list, 1 );
  free_list (mergeVCF_list, 1 );
  free_list (mergeVCFlist_list, 1 );
  
  cmdline_parser_release (&local_args_info);
  return (EXIT_FAILURE);
}

#ifndef CONFIG_FILE_LINE_SIZE
#define CONFIG_FILE_LINE_SIZE 2048
#endif
#define ADDITIONAL_ERROR " in configuration file "

#define CONFIG_FILE_LINE_BUFFER_SIZE (CONFIG_FILE_LINE_SIZE+3)
/* 3 is for "--" and "=" */

static int
_cmdline_parser_configfile (const char *filename, int *my_argc)
{
  FILE* file;
  char my_argv[CONFIG_FILE_LINE_BUFFER_SIZE+1];
  char linebuf[CONFIG_FILE_LINE_SIZE];
  int line_num = 0;
  int result = 0, equal;
  char *fopt, *farg;
  char *str_index;
  size_t len, next_token;
  char delimiter;

  if ((file = fopen(filename, "r")) == 0)
    {
      fprintf (stderr, "%s: Error opening configuration file '%s'\n",
               CMDLINE_PARSER_PACKAGE, filename);
      return EXIT_FAILURE;
    }

  while ((fgets(linebuf, CONFIG_FILE_LINE_SIZE, file)) != 0)
    {
      ++line_num;
      my_argv[0] = '\0';
      len = strlen(linebuf);
      if (len > (CONFIG_FILE_LINE_BUFFER_SIZE-1))
        {
          fprintf (stderr, "%s:%s:%d: Line too long in configuration file\n",
                   CMDLINE_PARSER_PACKAGE, filename, line_num);
          result = EXIT_FAILURE;
          break;
        }

      /* find first non-whitespace character in the line */
      next_token = strspn (linebuf, " \t\r\n");
      str_index  = linebuf + next_token;

      if ( str_index[0] == '\0' || str_index[0] == '#')
        continue; /* empty line or comment line is skipped */

      fopt = str_index;

      /* truncate fopt at the end of the first non-valid character */
      next_token = strcspn (fopt, " \t\r\n=");

      if (fopt[next_token] == '\0') /* the line is over */
        {
          farg  = 0;
          equal = 0;
          goto noarg;
        }

      /* remember if equal sign is present */
      equal = (fopt[next_token] == '=');
      fopt[next_token++] = '\0';

      /* advance pointers to the next token after the end of fopt */
      next_token += strspn (fopt + next_token, " \t\r\n");

      /* check for the presence of equal sign, and if so, skip it */
      if ( !equal )
        if ((equal = (fopt[next_token] == '=')))
          {
            next_token++;
            next_token += strspn (fopt + next_token, " \t\r\n");
          }
      str_index  += next_token;

      /* find argument */
      farg = str_index;
      if ( farg[0] == '\"' || farg[0] == '\'' )
        { /* quoted argument */
          str_index = strchr (++farg, str_index[0] ); /* skip opening quote */
          if (! str_index)
            {
              fprintf
                (stderr,
                 "%s:%s:%d: unterminated string in configuration file\n",
                 CMDLINE_PARSER_PACKAGE, filename, line_num);
              result = EXIT_FAILURE;
              break;
            }
        }
      else
        { /* read up the remaining part up to a delimiter */
          next_token = strcspn (farg, " \t\r\n#\'\"");
          str_index += next_token;
        }

      /* truncate farg at the delimiter and store it for further check */
      delimiter = *str_index, *str_index++ = '\0';

      /* everything but comment is illegal at the end of line */
      if (delimiter != '\0' && delimiter != '#')
        {
          str_index += strspn(str_index, " \t\r\n");
          if (*str_index != '\0' && *str_index != '#')
            {
              fprintf
                (stderr,
                 "%s:%s:%d: malformed string in configuration file\n",
                 CMDLINE_PARSER_PACKAGE, filename, line_num);
              result = EXIT_FAILURE;
              break;
            }
        }

    noarg:
      if (!strcmp(fopt,"include")) {
        if (farg && *farg) {
          result = _cmdline_parser_configfile(farg, my_argc);
        } else {
          fprintf(stderr, "%s:%s:%d: include requires a filename argument.\n",
                  CMDLINE_PARSER_PACKAGE, filename, line_num);
        }
        continue;
      }
      len = strlen(fopt);
      strcat (my_argv, len > 1 ? "--" : "-");
      strcat (my_argv, fopt);
      if (len > 1 && ((farg && *farg) || equal))
        strcat (my_argv, "=");
      if (farg && *farg)
        strcat (my_argv, farg);
      ++(*my_argc);

      cmd_line_list_tmp = (struct line_list *) malloc (sizeof (struct line_list));
      cmd_line_list_tmp->next = cmd_line_list;
      cmd_line_list = cmd_line_list_tmp;
      cmd_line_list->string_arg = gengetopt_strdup(my_argv);
    } /* while */

  if (file)
    fclose(file);
  return result;
}

int
cmdline_parser_configfile (
  const char *filename,
                           struct gengetopt_args_info *args_info,
                           int override, int initialize, int check_required)
{
  struct cmdline_parser_params params;

  params.override = override;
  params.initialize = initialize;
  params.check_required = check_required;
  params.check_ambiguity = 0;
  params.print_errors = 1;
  
  return cmdline_parser_config_file (filename, args_info, &params);
}

int
cmdline_parser_config_file (const char *filename,
                           struct gengetopt_args_info *args_info,
                           struct cmdline_parser_params *params)
{
  int i, result;
  int my_argc = 1;
  char **my_argv_arg;
  char *additional_error;

  /* store the program name */
  cmd_line_list_tmp = (struct line_list *) malloc (sizeof (struct line_list));
  cmd_line_list_tmp->next = cmd_line_list;
  cmd_line_list = cmd_line_list_tmp;
  cmd_line_list->string_arg = gengetopt_strdup (CMDLINE_PARSER_PACKAGE);

  result = _cmdline_parser_configfile(filename, &my_argc);

  if (result != EXIT_FAILURE) {
    my_argv_arg = (char **) malloc((my_argc+1) * sizeof(char *));
    cmd_line_list_tmp = cmd_line_list;

    for (i = my_argc - 1; i >= 0; --i) {
      my_argv_arg[i] = cmd_line_list_tmp->string_arg;
      cmd_line_list_tmp = cmd_line_list_tmp->next;
    }

    my_argv_arg[my_argc] = 0;

    additional_error = (char *)malloc(strlen(filename) + strlen(ADDITIONAL_ERROR) + 1);
    strcpy (additional_error, ADDITIONAL_ERROR);
    strcat (additional_error, filename);
    result =
      cmdline_parser_internal (my_argc, my_argv_arg, args_info,
                              params,
                              additional_error);

    free (additional_error);
    free (my_argv_arg);
  }

  free_cmd_list();
  if (result == EXIT_FAILURE)
    {
      cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}
